
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../05_clustering/">
      
      
        <link rel="next" href="../07_differential_expression/">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.4.7">
    
    
      
        <title>Cell type identification - Bioinformatics Education and Services at Tufts</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.4b4a2bd9.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.356b1318.min.css">
      
      
  
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
  
  <style>:root{--md-admonition-icon--note:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M1 7.775V2.75C1 1.784 1.784 1 2.75 1h5.025c.464 0 .91.184 1.238.513l6.25 6.25a1.75 1.75 0 0 1 0 2.474l-5.026 5.026a1.75 1.75 0 0 1-2.474 0l-6.25-6.25A1.752 1.752 0 0 1 1 7.775Zm1.5 0c0 .066.026.13.073.177l6.25 6.25a.25.25 0 0 0 .354 0l5.025-5.025a.25.25 0 0 0 0-.354l-6.25-6.25a.25.25 0 0 0-.177-.073H2.75a.25.25 0 0 0-.25.25ZM6 5a1 1 0 1 1 0 2 1 1 0 0 1 0-2Z"/></svg>');--md-admonition-icon--abstract:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M2.5 1.75v11.5c0 .138.112.25.25.25h3.17a.75.75 0 0 1 0 1.5H2.75A1.75 1.75 0 0 1 1 13.25V1.75C1 .784 1.784 0 2.75 0h8.5C12.216 0 13 .784 13 1.75v7.736a.75.75 0 0 1-1.5 0V1.75a.25.25 0 0 0-.25-.25h-8.5a.25.25 0 0 0-.25.25Zm13.274 9.537v-.001l-4.557 4.45a.75.75 0 0 1-1.055-.008l-1.943-1.95a.75.75 0 0 1 1.062-1.058l1.419 1.425 4.026-3.932a.75.75 0 1 1 1.048 1.074ZM4.75 4h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5ZM4 7.75A.75.75 0 0 1 4.75 7h2a.75.75 0 0 1 0 1.5h-2A.75.75 0 0 1 4 7.75Z"/></svg>');--md-admonition-icon--info:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M0 8a8 8 0 1 1 16 0A8 8 0 0 1 0 8Zm8-6.5a6.5 6.5 0 1 0 0 13 6.5 6.5 0 0 0 0-13ZM6.5 7.75A.75.75 0 0 1 7.25 7h1a.75.75 0 0 1 .75.75v2.75h.25a.75.75 0 0 1 0 1.5h-2a.75.75 0 0 1 0-1.5h.25v-2h-.25a.75.75 0 0 1-.75-.75ZM8 6a1 1 0 1 1 0-2 1 1 0 0 1 0 2Z"/></svg>');--md-admonition-icon--tip:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M3.499.75a.75.75 0 0 1 1.5 0v.996C5.9 2.903 6.793 3.65 7.662 4.376l.24.202c-.036-.694.055-1.422.426-2.163C9.1.873 10.794-.045 12.622.26 14.408.558 16 1.94 16 4.25c0 1.278-.954 2.575-2.44 2.734l.146.508.065.22c.203.701.412 1.455.476 2.226.142 1.707-.4 3.03-1.487 3.898C11.714 14.671 10.27 15 8.75 15h-6a.75.75 0 0 1 0-1.5h1.376a4.484 4.484 0 0 1-.563-1.191 3.835 3.835 0 0 1-.05-2.063 4.647 4.647 0 0 1-2.025-.293.75.75 0 0 1 .525-1.406c1.357.507 2.376-.006 2.698-.318l.009-.01a.747.747 0 0 1 1.06 0 .748.748 0 0 1-.012 1.074c-.912.92-.992 1.835-.768 2.586.221.74.745 1.337 1.196 1.621H8.75c1.343 0 2.398-.296 3.074-.836.635-.507 1.036-1.31.928-2.602-.05-.603-.216-1.224-.422-1.93l-.064-.221c-.12-.407-.246-.84-.353-1.29a2.425 2.425 0 0 1-.507-.441 3.075 3.075 0 0 1-.633-1.248.75.75 0 0 1 1.455-.364c.046.185.144.436.31.627.146.168.353.305.712.305.738 0 1.25-.615 1.25-1.25 0-1.47-.95-2.315-2.123-2.51-1.172-.196-2.227.387-2.706 1.345-.46.92-.27 1.774.019 3.062l.042.19a.884.884 0 0 1 .01.05c.348.443.666.949.94 1.553a.75.75 0 1 1-1.365.62c-.553-1.217-1.32-1.94-2.3-2.768L6.7 5.527c-.814-.68-1.75-1.462-2.692-2.619a3.737 3.737 0 0 0-1.023.88c-.406.495-.663 1.036-.722 1.508.116.122.306.21.591.239.388.038.797-.06 1.032-.19a.75.75 0 0 1 .728 1.31c-.515.287-1.23.439-1.906.373-.682-.067-1.473-.38-1.879-1.193L.75 5.677V5.5c0-.984.48-1.94 1.077-2.664.46-.559 1.05-1.055 1.673-1.353V.75Z"/></svg>');--md-admonition-icon--success:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z"/></svg>');--md-admonition-icon--question:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M0 8a8 8 0 1 1 16 0A8 8 0 0 1 0 8Zm8-6.5a6.5 6.5 0 1 0 0 13 6.5 6.5 0 0 0 0-13ZM6.92 6.085h.001a.749.749 0 1 1-1.342-.67c.169-.339.436-.701.849-.977C6.845 4.16 7.369 4 8 4a2.756 2.756 0 0 1 1.637.525c.503.377.863.965.863 1.725 0 .448-.115.83-.329 1.15-.205.307-.47.513-.692.662-.109.072-.22.138-.313.195l-.006.004a6.24 6.24 0 0 0-.26.16.952.952 0 0 0-.276.245.75.75 0 0 1-1.248-.832c.184-.264.42-.489.692-.661.103-.067.207-.132.313-.195l.007-.004c.1-.061.182-.11.258-.161a.969.969 0 0 0 .277-.245C8.96 6.514 9 6.427 9 6.25a.612.612 0 0 0-.262-.525A1.27 1.27 0 0 0 8 5.5c-.369 0-.595.09-.74.187a1.01 1.01 0 0 0-.34.398ZM9 11a1 1 0 1 1-2 0 1 1 0 0 1 2 0Z"/></svg>');--md-admonition-icon--warning:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M6.457 1.047c.659-1.234 2.427-1.234 3.086 0l6.082 11.378A1.75 1.75 0 0 1 14.082 15H1.918a1.75 1.75 0 0 1-1.543-2.575Zm1.763.707a.25.25 0 0 0-.44 0L1.698 13.132a.25.25 0 0 0 .22.368h12.164a.25.25 0 0 0 .22-.368Zm.53 3.996v2.5a.75.75 0 0 1-1.5 0v-2.5a.75.75 0 0 1 1.5 0ZM9 11a1 1 0 1 1-2 0 1 1 0 0 1 2 0Z"/></svg>');--md-admonition-icon--failure:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M2.344 2.343h-.001a8 8 0 0 1 11.314 11.314A8.002 8.002 0 0 1 .234 10.089a8 8 0 0 1 2.11-7.746Zm1.06 10.253a6.5 6.5 0 1 0 9.108-9.275 6.5 6.5 0 0 0-9.108 9.275ZM6.03 4.97 8 6.94l1.97-1.97a.749.749 0 0 1 1.275.326.749.749 0 0 1-.215.734L9.06 8l1.97 1.97a.749.749 0 0 1-.326 1.275.749.749 0 0 1-.734-.215L8 9.06l-1.97 1.97a.749.749 0 0 1-1.275-.326.749.749 0 0 1 .215-.734L6.94 8 4.97 6.03a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018Z"/></svg>');--md-admonition-icon--danger:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M9.504.43a1.516 1.516 0 0 1 2.437 1.713L10.415 5.5h2.123c1.57 0 2.346 1.909 1.22 3.004l-7.34 7.142a1.249 1.249 0 0 1-.871.354h-.302a1.25 1.25 0 0 1-1.157-1.723L5.633 10.5H3.462c-1.57 0-2.346-1.909-1.22-3.004L9.503.429Zm1.047 1.074L3.286 8.571A.25.25 0 0 0 3.462 9H6.75a.75.75 0 0 1 .694 1.034l-1.713 4.188 6.982-6.793A.25.25 0 0 0 12.538 7H9.25a.75.75 0 0 1-.683-1.06l2.008-4.418.003-.006a.036.036 0 0 0-.004-.009l-.006-.006-.008-.001c-.003 0-.006.002-.009.004Z"/></svg>');--md-admonition-icon--bug:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M4.72.22a.75.75 0 0 1 1.06 0l1 .999a3.488 3.488 0 0 1 2.441 0l.999-1a.748.748 0 0 1 1.265.332.75.75 0 0 1-.205.729l-.775.776c.616.63.995 1.493.995 2.444v.327c0 .1-.009.197-.025.292.408.14.764.392 1.029.722l1.968-.787a.75.75 0 0 1 .556 1.392L13 7.258V9h2.25a.75.75 0 0 1 0 1.5H13v.5c0 .409-.049.806-.141 1.186l2.17.868a.75.75 0 0 1-.557 1.392l-2.184-.873A4.997 4.997 0 0 1 8 16a4.997 4.997 0 0 1-4.288-2.427l-2.183.873a.75.75 0 0 1-.558-1.392l2.17-.868A5.036 5.036 0 0 1 3 11v-.5H.75a.75.75 0 0 1 0-1.5H3V7.258L.971 6.446a.75.75 0 0 1 .558-1.392l1.967.787c.265-.33.62-.583 1.03-.722a1.677 1.677 0 0 1-.026-.292V4.5c0-.951.38-1.814.995-2.444L4.72 1.28a.75.75 0 0 1 0-1.06Zm.53 6.28a.75.75 0 0 0-.75.75V11a3.5 3.5 0 1 0 7 0V7.25a.75.75 0 0 0-.75-.75ZM6.173 5h3.654A.172.172 0 0 0 10 4.827V4.5a2 2 0 1 0-4 0v.327c0 .096.077.173.173.173Z"/></svg>');--md-admonition-icon--example:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M5 5.782V2.5h-.25a.75.75 0 0 1 0-1.5h6.5a.75.75 0 0 1 0 1.5H11v3.282l3.666 5.76C15.619 13.04 14.543 15 12.767 15H3.233c-1.776 0-2.852-1.96-1.899-3.458Zm-2.4 6.565a.75.75 0 0 0 .633 1.153h9.534a.75.75 0 0 0 .633-1.153L12.225 10.5h-8.45ZM9.5 2.5h-3V6c0 .143-.04.283-.117.403L4.73 9h6.54L9.617 6.403A.746.746 0 0 1 9.5 6Z"/></svg>');--md-admonition-icon--prerequisite:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M5 5.782V2.5h-.25a.75.75 0 0 1 0-1.5h6.5a.75.75 0 0 1 0 1.5H11v3.282l3.666 5.76C15.619 13.04 14.543 15 12.767 15H3.233c-1.776 0-2.852-1.96-1.899-3.458Zm-2.4 6.565a.75.75 0 0 0 .633 1.153h9.534a.75.75 0 0 0 .633-1.153L12.225 10.5h-8.45ZM9.5 2.5h-3V6c0 .143-.04.283-.117.403L4.73 9h6.54L9.617 6.403A.746.746 0 0 1 9.5 6Z"/></svg>');--md-admonition-icon--quote:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M1.75 2.5h10.5a.75.75 0 0 1 0 1.5H1.75a.75.75 0 0 1 0-1.5Zm4 5h8.5a.75.75 0 0 1 0 1.5h-8.5a.75.75 0 0 1 0-1.5Zm0 5h8.5a.75.75 0 0 1 0 1.5h-8.5a.75.75 0 0 1 0-1.5ZM2.5 7.75v6a.75.75 0 0 1-1.5 0v-6a.75.75 0 0 1 1.5 0Z"/></svg>');}</style>



    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="blue">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#cell-type-identification" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Bioinformatics Education and Services at Tufts" class="md-header__button md-logo" aria-label="Bioinformatics Education and Services at Tufts" data-md-component="logo">
      
  <img src="../../images/logo1.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Bioinformatics Education and Services at Tufts
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Cell type identification
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="blue"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="blue"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
</form>
      
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/best-tufts/intro_to_single_cell_rnaseq" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../00_background/" class="md-tabs__link">
          
  
  Workshop Materials

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../.." class="md-tabs__link">
          
  
  All Workshops

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Bioinformatics Education and Services at Tufts" class="md-nav__button md-logo" aria-label="Bioinformatics Education and Services at Tufts" data-md-component="logo">
      
  <img src="../../images/logo1.svg" alt="logo">

    </a>
    Bioinformatics Education and Services at Tufts
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/best-tufts/intro_to_single_cell_rnaseq" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
    
  
  
    
    
    
    
    
      
      
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_1" checked>
        
          
          <label class="md-nav__link" for="__nav_1" id="__nav_1_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Workshop Materials
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_1">
            <span class="md-nav__icon md-icon"></span>
            Workshop Materials
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../00_background/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Schedule
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../01_setup/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Set up
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../02_formatting/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Formatting
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../03_quality_control/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quality Control
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../04_integration/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Integration
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../05_clustering/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Clustering
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Cell Type Identification
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Cell Type Identification
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#singler-correlation-method" class="md-nav__link">
    SingleR Correlation Method
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#seurat-integration-mapping" class="md-nav__link">
    Seurat Integration Mapping
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../07_differential_expression/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Differential Expression
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
    
    
    
    
    
      
      
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="">
            
  
  <span class="md-ellipsis">
    All Workshops
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            All Workshops
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    All Workshops
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#singler-correlation-method" class="md-nav__link">
    SingleR Correlation Method
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#seurat-integration-mapping" class="md-nav__link">
    Seurat Integration Mapping
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
    <a href="https://github.com/best-tufts/intro_to_single_cell_rnaseq/blob/main/docs/intro_to_single_cell_rnaseq/06_cell_type_identification.md" title="Edit this page" class="md-content__button md-icon">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4v-2m10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1 2.1 2.1Z"/></svg>
    </a>
  
  


<h1 id="cell-type-identification">Cell-type Identification</h1>
<p>In this section, we'll demonstrate two automated methods to label cells in our dataset using reference datasets with known cell labels. More information on these methods is presented in <a href="../slides/day2_temp.pdf">Lecture Part II</a>.</p>
<ul>
<li><a href="https://bioconductor.org/packages/devel/bioc/vignettes/SingleR/inst/doc/SingleR.html">SingleR</a> method, which uses correlation of gene expression. This method can use both single-cell and bulk RNAseq reference datasets. </li>
<li><a href="https://satijalab.org/seurat/articles/integration_mapping.html">Seurat Integration Mapping</a> which applies integration between a labeled, reference single-cell RNAseq dataset and our query dataset</li>
</ul>
<p>To start, we set our library path on the HPC cluster:
<div class="highlight"><pre><span></span><code><span class="n">LIB</span><span class="o">=</span><span class="s">&#39;/cluster/tufts/hpc/tools/R/4.0.0/&#39;</span>
<span class="nf">.libPaths</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;&quot;</span><span class="p">,</span><span class="n">LIB</span><span class="p">))</span>
</code></pre></div></p>
<p>We require three new packages:
- <a href="https://bioconductor.org/packages/release/bioc/html/SingleR.html">Singler</a>
- <a href="https://bioconductor.org/packages/release/data/experiment/html/celldex.html">celldex</a>: A collection of reference expression datasets with cell type labels 
- <a href="https://cran.r-project.org/web/packages/pheatmap/index.html">pheatmap</a> A package to make "pretty" heatmaps.</p>
<div class="highlight"><pre><span></span><code><span class="nf">suppressPackageStartupMessages</span><span class="p">({</span>
<span class="w">  </span><span class="nf">library</span><span class="p">(</span><span class="n">Seurat</span><span class="p">)</span>
<span class="w">  </span><span class="nf">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
<span class="w">  </span><span class="nf">library</span><span class="p">(</span><span class="n">cowplot</span><span class="p">)</span>
<span class="w">  </span><span class="nf">library</span><span class="p">(</span><span class="n">SingleR</span><span class="p">)</span>
<span class="w">  </span><span class="nf">library</span><span class="p">(</span><span class="n">celldex</span><span class="p">)</span>
<span class="w">  </span><span class="nf">library</span><span class="p">(</span><span class="n">pheatmap</span><span class="p">)</span>
<span class="p">})</span>
</code></pre></div>
<p>Set the base dir:
<div class="highlight"><pre><span></span><code><span class="n">baseDir</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;~/intro_to_scrnaseq/&quot;</span>
</code></pre></div></p>
<p>We begin by loading a pre-processed version of our integrated, clustered samples from the <code>data</code> folder.
<div class="highlight"><pre><span></span><code><span class="n">integ_seurat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">readRDS</span><span class="p">(</span><span class="nf">file.path</span><span class="p">(</span><span class="n">baseDir</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;data/clustered_seurat.rds&quot;</span><span class="p">))</span>
</code></pre></div></p>
<p>When we left off, we had integrated our unstimulated (<code>ctrl</code>) and interferon beta stimulated (<code>stim</code>) PBMC samples and clustered the resulting dataset at resolution 0.4. We can visualize the cells both using either cluster or sample labels:
<div class="highlight"><pre><span></span><code><span class="nf">Idents</span><span class="p">(</span><span class="n">object</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">integ_seurat</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;integrated_snn_res.0.4&quot;</span>
<span class="n">p1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">DimPlot</span><span class="p">(</span><span class="n">integ_seurat</span><span class="p">,</span><span class="w"> </span><span class="n">label</span><span class="o">=</span><span class="bp">T</span><span class="p">)</span>

<span class="nf">Idents</span><span class="p">(</span><span class="n">object</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">integ_seurat</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;sample&quot;</span>
<span class="n">p2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">DimPlot</span><span class="p">(</span><span class="n">integ_seurat</span><span class="p">)</span>

<span class="nf">plot_grid</span><span class="p">(</span><span class="n">p1</span><span class="p">,</span><span class="w"> </span><span class="n">p2</span><span class="p">)</span>
</code></pre></div>
<img alt="" src="../images/cluster_ident_grid_umap.png" /></p>
<h2 id="singler-correlation-method">SingleR Correlation Method</h2>
<p>First, we'll use <a href="https://github.com/LTLA/SingleR">SingleR</a> along with a reference database of expression profiles of known cell types in order to identify our cells and clusters. As mentioned in the lecture, this method measures the correlation of overall gene expression between cells in a reference database with cells in the query dataset in order to label cells.</p>
<p>To start, we'll use a database of bulk RNAseq profiles of Human pure cell-types called the Human Primary Cell Type Atlas.  This dataset along with several others is available through the <a href="https://rdrr.io/github/LTLA/celldex/man/HumanPrimaryCellAtlasData.html">celldex</a> R library. A database like this is useful as a starting point to determine which broad cell types your sample has, either because there is no existing single-cell reference for your sample-type, or because you want to make sure that your sample-type contains the expected cell types before using a more specific single-cell reference. </p>
<p>To load, type the following, and type <code>yes</code> when prompted to create the <code>ExperimentHub</code> directory in a hidden <code>.cache</code> directory in your home folder.
<div class="highlight"><pre><span></span><code><span class="n">hpca</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">HumanPrimaryCellAtlasData</span><span class="p">()</span>
</code></pre></div>
The HPCA object has a data type called <code>Summarized Experiment</code> which allows one to store numeric data matrices in <code>assays</code> along with a data.frame providing annotation columns for sample in the data. </p>
<p><img alt="" src="../images/hpca_graphic.png" /></p>
<p>We can access the sample annotations in the <code>@colData</code> slot, which show the sample cell-types. We can access the log-transformed expression data in the <code>@assay</code> slot.
<div class="highlight"><pre><span></span><code><span class="nf">head</span><span class="p">(</span><span class="n">hpca</span><span class="o">@</span><span class="n">colData</span><span class="p">)</span>
<span class="nf">head</span><span class="p">(</span><span class="n">hpca</span><span class="o">@</span><span class="n">assays</span><span class="o">@</span><span class="n">data</span><span class="o">$</span><span class="n">logcounts</span><span class="p">)</span>
</code></pre></div></p>
<p>Well use the coarse-grained cell-type labels, <code>label.main</code>, column of the metadata, which has the following cell-types:
<div class="highlight"><pre><span></span><code><span class="nf">unique</span><span class="p">(</span><span class="n">hpca</span><span class="o">$</span><span class="n">label.main</span><span class="p">)</span>
</code></pre></div>
<img alt="" src="../images/unique_hpca.png" /></p>
<p>The data to be labeled is input into SingleR as a log-transformed, normalized count matrix, which we can extract from the <code>RNA</code> assay our <code>integ_seurat</code> object:
<div class="highlight"><pre><span></span><code><span class="n">query_counts</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">integ_seurat</span><span class="o">@</span><span class="n">assays</span><span class="o">$</span><span class="n">RNA</span><span class="o">@</span><span class="n">data</span>
</code></pre></div></p>
<p><code>SingleR</code> can be run both on the cluster level and the individual cell level. For cluster-level annotation, the average expression profile of each cluster is used and a single label is generated. This is much faster to run, since we have many fewer clusters than cells, so we'll start here.</p>
<div class="highlight"><pre><span></span><code><span class="n">query_clusters</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">integ_seurat</span><span class="o">@</span><span class="n">meta.data</span><span class="o">$</span><span class="n">integrated_snn_res.0.4</span>
</code></pre></div>
<p>The following command runs SingleR on the cluster level, without pruning, which should take only a few seconds.
<div class="highlight"><pre><span></span><code><span class="n">pred_cluster</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">SingleR</span><span class="p">(</span><span class="n">test</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">query_counts</span><span class="p">,</span>
<span class="w">                        </span><span class="n">clusters</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">query_clusters</span><span class="p">,</span>
<span class="w">                        </span><span class="n">ref</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">hpca</span><span class="p">,</span>
<span class="w">                        </span><span class="n">labels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">hpca</span><span class="o">$</span><span class="n">label.main</span><span class="p">,</span><span class="w"> </span>
<span class="w">                        </span><span class="n">prune</span><span class="o">=</span><span class="bp">F</span><span class="p">)</span>
</code></pre></div></p>
<p>Save the results:
<div class="highlight"><pre><span></span><code><span class="nf">saveRDS</span><span class="p">(</span><span class="n">pred_cluster</span><span class="p">,</span><span class="w"> </span><span class="nf">file.path</span><span class="p">(</span><span class="n">baseDir</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;results/singler_hpca_cluster_res0.4.rds&quot;</span><span class="p">))</span>
</code></pre></div></p>
<p>We can view the results, which contain a score for every cell type plus the final label:
<div class="highlight"><pre><span></span><code><span class="nf">view</span><span class="p">(</span><span class="n">pred_cluster</span><span class="p">)</span>
</code></pre></div></p>
<p>Select the columns that start with <code>score</code> to plot as a heatmap and the <code>label</code> column to annotate the heatmap:
<div class="highlight"><pre><span></span><code><span class="n">scores</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">data.frame</span><span class="p">(</span><span class="n">pred_cluster</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="n">dplyr</span><span class="o">::</span><span class="nf">select</span><span class="p">(</span><span class="nf">starts_with</span><span class="p">(</span><span class="s">&quot;scores&quot;</span><span class="p">))</span><span class="w"> </span>

<span class="n">labels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">data.frame</span><span class="p">(</span><span class="n">pred_cluster</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="n">dplyr</span><span class="o">::</span><span class="nf">select</span><span class="p">(</span><span class="s">&quot;labels&quot;</span><span class="p">)</span>
</code></pre></div></p>
<p>The scores and chosen labels can be plotted using <code>pheatmap</code>:
<div class="highlight"><pre><span></span><code><span class="nf">pheatmap</span><span class="p">(</span><span class="n">scores</span><span class="p">,</span>
<span class="w">         </span><span class="n">annotation_row</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">labels</span><span class="p">)</span><span class="w"> </span>
</code></pre></div>
<img alt="" src="../images/pred_cluster_pheatmap.png" /></p>
<p>Four cell-types were found in our data: B_cell, Monocyte, NK_cell and T_cell. However, some clusters like 10, 11, 12, do not appear to have a good match for any reference cell type.  </p>
<p>Let's add the labels make a named list with new names:
<div class="highlight"><pre><span></span><code><span class="n">new_names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pred_cluster</span><span class="o">$</span><span class="n">labels</span>
<span class="nf">names</span><span class="p">(</span><span class="n">new_names</span><span class="p">)</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rownames</span><span class="p">(</span><span class="n">pred_cluster</span><span class="p">)</span>
<span class="n">new_names</span>
</code></pre></div>
<img alt="" src="../images/cluster_label.png" /></p>
<p>Set the identities to the clusters found at resolution 0.4, rename the identities, and add the names to the Seurat metadata:
<div class="highlight"><pre><span></span><code><span class="nf">Idents</span><span class="p">(</span><span class="n">object</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">integ_seurat</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;integrated_snn_res.0.4&quot;</span>
<span class="n">integ_seurat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">RenameIdents</span><span class="p">(</span><span class="n">integ_seurat</span><span class="p">,</span><span class="w"> </span>
<span class="w">                                 </span><span class="n">new_names</span><span class="p">)</span>
<span class="n">integ_seurat</span><span class="o">$</span><span class="n">singler_cluster_labels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">Idents</span><span class="p">(</span><span class="n">integ_seurat</span><span class="p">)</span>
</code></pre></div></p>
<p>Plot the labeled clusters:
<div class="highlight"><pre><span></span><code><span class="nf">Idents</span><span class="p">(</span><span class="n">integ_seurat</span><span class="p">)</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;singler_cluster_labels&quot;</span>
<span class="nf">DimPlot</span><span class="p">(</span><span class="n">integ_seurat</span><span class="p">,</span><span class="w"> </span>
<span class="w">        </span><span class="n">label</span><span class="o">=</span><span class="bp">T</span><span class="p">)</span>
</code></pre></div>
<img alt="" src="../images/pred_cluster_label.png" /></p>
<p>Running <code>SingleR</code> on the cell level uses the same method but eliminates the <code>clusters</code> argument from the command and enables the default pruning process. It will take ~20 minutes to run, so we'll run it as a batch job and not interactively (do not run the below code in Rstudio).</p>
<p><div class="highlight"><pre><span></span><code># DON&#39;T RUN!
pred_cell &lt;- SingleR(test = query_counts,
                      ref = hpca,
                      assay.type.test=&quot;logcounts&quot;,
                      labels = hpca$label.main)
</code></pre></div>
Running R scripts as batch job is convenient and does not require the Rstudio session to remain connected while the job runs. To navigate to our scripts directory and open <code>06_singler_cell.R</code>. This file contains the key steps above but runs <code>SingleR</code> on the cell level.</p>
<p>To run it, we use an <code>SBATCH</code> file that is interpreted by the cluster job scheduler called <a href="https://tufts.app.box.com/file/780443358198?v=Pax-User-Guide">slurm</a>. Open the file <code>06_run_singler_cell.sh</code> in the <code>scripts</code> directory. </p>
<p>The file contains a header which specifies the resources that the job will need so that an appropriate compute node can be allocated. Header lines start with <code>#SBATCH</code>. The body of the script specifies the code to be run once the job is started on a compute node. </p>
<div class="highlight"><pre><span></span><code><span class="ch">#!/bin/bash                       # Use the bash shell interpreter</span>
<span class="c1">#SBATCH -J run_singler            # Give the job a name</span>
<span class="c1">#SBATCH --time=2:00:00            # Request 2 hours</span>
<span class="c1">#SBATCH -n 1                      # Request 1 core</span>
<span class="c1">#SBATCH -N 1                      # Request 1 node</span>
<span class="c1">#SBATCH --mem=10Gb                # Request 10 Gb</span>
<span class="c1">#SBATCH --output=%j.out           # Write the job output to a file prefixed by the job number</span>
<span class="c1">#SBATCH --error=%j.err            # Write the job error to a file prefixed by the job number</span>

module<span class="w"> </span>purge<span class="w">                      </span><span class="c1"># Remove loaded modules</span>
module<span class="w"> </span>load<span class="w"> </span>R/4.0.0<span class="w">               </span><span class="c1"># Load R module</span>

Rscript<span class="w"> </span>--no-save<span class="w"> </span>06_singler_cell.R<span class="w">  </span><span class="c1"># Run the script </span>
</code></pre></div>
<p>To run the script:
- Click on <code>Terminal</code> next to <code>Console</code> in the bottom portion of the <code>Rstudio</code> application
- Change to our <code>scripts</code> directory by typing <code>cd ~/intro_to_scrnaseq/scripts</code>
- Type <code>sbatch 06_run_singler_cell.sh</code> and press enter. 
- Your job will be given a number by <code>slurm</code> and placed in the queue.
- To check the status of your job, type <code>squeue -u tufts-username</code> and you will see your job status. </p>
<details class="question">
<summary>Is your job submitted successfully?</summary>
<ul>
<li>Yes (put up a green check mark in zoom)</li>
<li>No (raise hand in zoom)</li>
</ul>
</details>
<p>In the meantime, let's load the pre-processed cell-level labels.
<div class="highlight"><pre><span></span><code><span class="n">pred_cell</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">readRDS</span><span class="p">(</span><span class="nf">file.path</span><span class="p">(</span><span class="n">baseDir</span><span class="p">,</span><span class="s">&quot;data/singler_hpca_cell.rds&quot;</span><span class="p">))</span>
<span class="nf">head</span><span class="p">(</span><span class="n">pred_cell</span><span class="p">)</span>
</code></pre></div>
<img alt="" src="../images/head_pred_cell.png" /></p>
<p>The cell-level labels have one row for every cell. The table contains scores for each reference label as well as both a <code>labels</code> and a <code>pruned.labels</code> column. We can directly add the <code>pruned.labels</code> to the Seurat object metadata using the <code>AddMetaData</code> function. </p>
<div class="highlight"><pre><span></span><code><span class="n">integ_seurat</span><span class="w">  </span><span class="o">=</span><span class="w"> </span><span class="nf">AddMetaData</span><span class="p">(</span><span class="n">integ_seurat</span><span class="p">,</span>
<span class="w">                                 </span><span class="n">pred_cell</span><span class="o">$</span><span class="n">pruned.labels</span><span class="p">,</span>
<span class="w">                                 </span><span class="n">col.name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;singler_cell_labels&quot;</span><span class="p">)</span>
</code></pre></div>
<p>Assign the idents and make a plot:
<div class="highlight"><pre><span></span><code><span class="nf">Idents</span><span class="p">(</span><span class="n">integ_seurat</span><span class="p">)</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;singler_cell_labels&quot;</span>
<span class="nf">DimPlot</span><span class="p">(</span><span class="n">integ_seurat</span><span class="p">,</span><span class="w"> </span>
<span class="w">        </span><span class="n">label</span><span class="o">=</span><span class="bp">T</span><span class="p">)</span>
</code></pre></div>
<img alt="" src="../images/pred_cell_label.png" /></p>
<p>Some clusters contain a mix of cell labels. We can view the breakdown of cell-type labels per cluster using pheatmap:
<div class="highlight"><pre><span></span><code><span class="n">tab</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">table</span><span class="p">(</span><span class="n">cluster</span><span class="o">=</span><span class="n">integ_seurat</span><span class="o">$</span><span class="n">integrated_snn_res.0.4</span><span class="p">,</span>
<span class="w">             </span><span class="n">label</span><span class="o">=</span><span class="n">pred_cell</span><span class="o">$</span><span class="n">labels</span><span class="p">)</span>

<span class="c1"># divide by the total number of cells in each cluster</span>
<span class="n">tab</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">tab</span><span class="o">/</span><span class="nf">rowSums</span><span class="p">(</span><span class="n">tab</span><span class="p">)</span>

<span class="nf">pheatmap</span><span class="p">(</span><span class="n">tab</span><span class="p">)</span><span class="w"> </span>
</code></pre></div>
<img alt="" src="../images/pred_cell_heatmap.png" /></p>
<details class="question">
<summary>Are the dominant cell-level labels the same as the cluster-level labels?</summary>
<ul>
<li>Add a row annotation of the cluster-level labels in order to compare visually</li>
</ul>
</details>
<p>Some clusters appear to have a mix of cells, which may indicate that they contain a type of cell not in our reference database. This is expected since we've used a very general database. Next we'll use a single-cell RNAseq dataset that we expect will have a better match to our data.</p>
<h2 id="seurat-integration-mapping">Seurat Integration Mapping</h2>
<p>The next reference dataset is a PBMC dataset <a href="https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz">available from 10X Genomics</a> which has been pre-processed through the Seurat pipeline. Load the data:
<div class="highlight"><pre><span></span><code><span class="n">pbmc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">readRDS</span><span class="p">(</span><span class="nf">file.path</span><span class="p">(</span><span class="n">baseDir</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;data/pbmc_reference.rds&quot;</span><span class="p">))</span>
</code></pre></div></p>
<p>The cell-type label is in the <code>seurat_annotation</code> metadata column. Set the cell identities to this column and plot:
<div class="highlight"><pre><span></span><code><span class="nf">Idents</span><span class="p">(</span><span class="n">pbmc</span><span class="p">)</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;seurat_annotations&quot;</span>
<span class="nf">DimPlot</span><span class="p">(</span><span class="n">pbmc</span><span class="p">,</span><span class="w"> </span><span class="n">label</span><span class="o">=</span><span class="bp">T</span><span class="p">)</span>
</code></pre></div></p>
<p><img alt="" src="../images/pbmc_reference_label.png" /></p>
<p>We see there are more specific T-cell and Monocyte cell subset labels. </p>
<p>The first step is to find the transfer anchors using the function <code>FindTransferAnchors</code> with the following code. This step similar to the <code>FindIntegrationAnchors</code> function in the integration step, except the PCA is performed only on the reference dataset and the query dataset is projected onto the reference learned PCA structure (<code>reduction = "pcaproject"</code> and <code>project.query = FALSE</code> by default). This allows us to potentially use a reference dataset with many more cell types than our query dataset. </p>
<div class="highlight"><pre><span></span><code><span class="n">anchors</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">FindTransferAnchors</span><span class="p">(</span><span class="n">reference</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pbmc</span><span class="p">,</span><span class="w"> </span>
<span class="w">                                   </span><span class="n">query</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">integ_seurat</span><span class="p">)</span>
</code></pre></div>
<p>Next, we'll transfer the reference cell-type labels to the query using <code>TransferData</code>. This step uses the proximity of each query cells to each anchors in order to label query cells.
<div class="highlight"><pre><span></span><code><span class="n">predictions</span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">TransferData</span><span class="p">(</span><span class="n">anchorset</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">anchors</span><span class="p">,</span><span class="w"> </span>
<span class="w">                                </span><span class="n">refdata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pbmc</span><span class="o">$</span><span class="n">seurat_annotations</span><span class="p">)</span>
</code></pre></div></p>
<p>Add the predicted id to the metadata:
<div class="highlight"><pre><span></span><code><span class="n">integ_seurat</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">AddMetaData</span><span class="p">(</span><span class="n">integ_seurat</span><span class="p">,</span><span class="w"> </span>
<span class="w">                                 </span><span class="n">metadata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">predictions</span><span class="o">$</span><span class="n">predicted.id</span><span class="p">,</span>
<span class="w">                                 </span><span class="n">col.name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;seurat_labels&quot;</span><span class="p">)</span>
</code></pre></div></p>
<p>Set the identities to our newly added column and plot:
<div class="highlight"><pre><span></span><code><span class="nf">Idents</span><span class="p">(</span><span class="n">integ_seurat</span><span class="p">)</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;seurat_labels&quot;</span>
<span class="nf">DimPlot</span><span class="p">(</span><span class="n">integ_seurat</span><span class="p">,</span><span class="w"> </span>
<span class="w">            </span><span class="n">label</span><span class="o">=</span><span class="bp">T</span><span class="w"> </span><span class="p">)</span>
</code></pre></div></p>
<p><img alt="" src="../images/integrate_label.png" /></p>
<p>We can view the breakdown per cluster as a heatmap:
<div class="highlight"><pre><span></span><code><span class="n">tab</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">table</span><span class="p">(</span><span class="n">cluster</span><span class="o">=</span><span class="n">integ_seurat</span><span class="o">$</span><span class="n">integrated_snn_res.0.4</span><span class="p">,</span>
<span class="w">             </span><span class="n">label</span><span class="o">=</span><span class="n">integ_seurat</span><span class="o">$</span><span class="n">seurat_labels</span><span class="p">)</span>

<span class="c1"># divide by the total number of cells in each cluster</span>
<span class="n">tab</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">tab</span><span class="o">/</span><span class="nf">rowSums</span><span class="p">(</span><span class="n">tab</span><span class="p">)</span>

<span class="nf">pheatmap</span><span class="p">(</span><span class="n">tab</span><span class="p">)</span><span class="w"> </span>
</code></pre></div></p>
<p><img alt="" src="../images/integrate_heatmap.png" /></p>
<p>Cluster 11 cells are labeled predominantly Dendritic Cells. </p>
<p>Finally, we save the labeled object:
<div class="highlight"><pre><span></span><code><span class="nf">saveRDS</span><span class="p">(</span><span class="n">integ_seurat</span><span class="p">,</span><span class="w"> </span><span class="nf">file.path</span><span class="p">(</span><span class="n">baseDir</span><span class="p">,</span><span class="s">&quot;results/labeled_seurat.rds&quot;</span><span class="p">))</span>
</code></pre></div></p>
<p>In the next section we'll look at a type of Differentially Expressed Gene known as a Marker Gene, which can help to confirm cell-type labels by confirming expression of known cell-type specific genes.</p>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../05_clustering/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Clustering">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                Clustering
              </div>
            </div>
          </a>
        
        
          
          <a href="../07_differential_expression/" class="md-footer__link md-footer__link--next" aria-label="Next: Differential Expression">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                Differential Expression
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tabs", "navigation.tabs.sticky", "navigation.sections", "content.action.edit", "navigation.footer", "content.code.copy"], "search": "../../assets/javascripts/workers/search.f886a092.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.aecac24b.min.js"></script>
      
        <script src="../../javascripts/mathjax.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
        <script src="https://unpkg.com/tablesort@5.3.0/dist/tablesort.min.js"></script>
      
        <script src="../../javascripts/tablesort.js"></script>
      
    
  </body>
</html>