
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../01_setup/">
      
      
        <link rel="next" href="../03_quality_control/">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.4.7">
    
    
      
        <title>Formatting - Bioinformatics Education and Services at Tufts</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.4b4a2bd9.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.356b1318.min.css">
      
      
  
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
  
  <style>:root{--md-admonition-icon--note:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M1 7.775V2.75C1 1.784 1.784 1 2.75 1h5.025c.464 0 .91.184 1.238.513l6.25 6.25a1.75 1.75 0 0 1 0 2.474l-5.026 5.026a1.75 1.75 0 0 1-2.474 0l-6.25-6.25A1.752 1.752 0 0 1 1 7.775Zm1.5 0c0 .066.026.13.073.177l6.25 6.25a.25.25 0 0 0 .354 0l5.025-5.025a.25.25 0 0 0 0-.354l-6.25-6.25a.25.25 0 0 0-.177-.073H2.75a.25.25 0 0 0-.25.25ZM6 5a1 1 0 1 1 0 2 1 1 0 0 1 0-2Z"/></svg>');--md-admonition-icon--abstract:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M2.5 1.75v11.5c0 .138.112.25.25.25h3.17a.75.75 0 0 1 0 1.5H2.75A1.75 1.75 0 0 1 1 13.25V1.75C1 .784 1.784 0 2.75 0h8.5C12.216 0 13 .784 13 1.75v7.736a.75.75 0 0 1-1.5 0V1.75a.25.25 0 0 0-.25-.25h-8.5a.25.25 0 0 0-.25.25Zm13.274 9.537v-.001l-4.557 4.45a.75.75 0 0 1-1.055-.008l-1.943-1.95a.75.75 0 0 1 1.062-1.058l1.419 1.425 4.026-3.932a.75.75 0 1 1 1.048 1.074ZM4.75 4h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5ZM4 7.75A.75.75 0 0 1 4.75 7h2a.75.75 0 0 1 0 1.5h-2A.75.75 0 0 1 4 7.75Z"/></svg>');--md-admonition-icon--info:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M0 8a8 8 0 1 1 16 0A8 8 0 0 1 0 8Zm8-6.5a6.5 6.5 0 1 0 0 13 6.5 6.5 0 0 0 0-13ZM6.5 7.75A.75.75 0 0 1 7.25 7h1a.75.75 0 0 1 .75.75v2.75h.25a.75.75 0 0 1 0 1.5h-2a.75.75 0 0 1 0-1.5h.25v-2h-.25a.75.75 0 0 1-.75-.75ZM8 6a1 1 0 1 1 0-2 1 1 0 0 1 0 2Z"/></svg>');--md-admonition-icon--tip:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M3.499.75a.75.75 0 0 1 1.5 0v.996C5.9 2.903 6.793 3.65 7.662 4.376l.24.202c-.036-.694.055-1.422.426-2.163C9.1.873 10.794-.045 12.622.26 14.408.558 16 1.94 16 4.25c0 1.278-.954 2.575-2.44 2.734l.146.508.065.22c.203.701.412 1.455.476 2.226.142 1.707-.4 3.03-1.487 3.898C11.714 14.671 10.27 15 8.75 15h-6a.75.75 0 0 1 0-1.5h1.376a4.484 4.484 0 0 1-.563-1.191 3.835 3.835 0 0 1-.05-2.063 4.647 4.647 0 0 1-2.025-.293.75.75 0 0 1 .525-1.406c1.357.507 2.376-.006 2.698-.318l.009-.01a.747.747 0 0 1 1.06 0 .748.748 0 0 1-.012 1.074c-.912.92-.992 1.835-.768 2.586.221.74.745 1.337 1.196 1.621H8.75c1.343 0 2.398-.296 3.074-.836.635-.507 1.036-1.31.928-2.602-.05-.603-.216-1.224-.422-1.93l-.064-.221c-.12-.407-.246-.84-.353-1.29a2.425 2.425 0 0 1-.507-.441 3.075 3.075 0 0 1-.633-1.248.75.75 0 0 1 1.455-.364c.046.185.144.436.31.627.146.168.353.305.712.305.738 0 1.25-.615 1.25-1.25 0-1.47-.95-2.315-2.123-2.51-1.172-.196-2.227.387-2.706 1.345-.46.92-.27 1.774.019 3.062l.042.19a.884.884 0 0 1 .01.05c.348.443.666.949.94 1.553a.75.75 0 1 1-1.365.62c-.553-1.217-1.32-1.94-2.3-2.768L6.7 5.527c-.814-.68-1.75-1.462-2.692-2.619a3.737 3.737 0 0 0-1.023.88c-.406.495-.663 1.036-.722 1.508.116.122.306.21.591.239.388.038.797-.06 1.032-.19a.75.75 0 0 1 .728 1.31c-.515.287-1.23.439-1.906.373-.682-.067-1.473-.38-1.879-1.193L.75 5.677V5.5c0-.984.48-1.94 1.077-2.664.46-.559 1.05-1.055 1.673-1.353V.75Z"/></svg>');--md-admonition-icon--success:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z"/></svg>');--md-admonition-icon--question:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M0 8a8 8 0 1 1 16 0A8 8 0 0 1 0 8Zm8-6.5a6.5 6.5 0 1 0 0 13 6.5 6.5 0 0 0 0-13ZM6.92 6.085h.001a.749.749 0 1 1-1.342-.67c.169-.339.436-.701.849-.977C6.845 4.16 7.369 4 8 4a2.756 2.756 0 0 1 1.637.525c.503.377.863.965.863 1.725 0 .448-.115.83-.329 1.15-.205.307-.47.513-.692.662-.109.072-.22.138-.313.195l-.006.004a6.24 6.24 0 0 0-.26.16.952.952 0 0 0-.276.245.75.75 0 0 1-1.248-.832c.184-.264.42-.489.692-.661.103-.067.207-.132.313-.195l.007-.004c.1-.061.182-.11.258-.161a.969.969 0 0 0 .277-.245C8.96 6.514 9 6.427 9 6.25a.612.612 0 0 0-.262-.525A1.27 1.27 0 0 0 8 5.5c-.369 0-.595.09-.74.187a1.01 1.01 0 0 0-.34.398ZM9 11a1 1 0 1 1-2 0 1 1 0 0 1 2 0Z"/></svg>');--md-admonition-icon--warning:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M6.457 1.047c.659-1.234 2.427-1.234 3.086 0l6.082 11.378A1.75 1.75 0 0 1 14.082 15H1.918a1.75 1.75 0 0 1-1.543-2.575Zm1.763.707a.25.25 0 0 0-.44 0L1.698 13.132a.25.25 0 0 0 .22.368h12.164a.25.25 0 0 0 .22-.368Zm.53 3.996v2.5a.75.75 0 0 1-1.5 0v-2.5a.75.75 0 0 1 1.5 0ZM9 11a1 1 0 1 1-2 0 1 1 0 0 1 2 0Z"/></svg>');--md-admonition-icon--failure:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M2.344 2.343h-.001a8 8 0 0 1 11.314 11.314A8.002 8.002 0 0 1 .234 10.089a8 8 0 0 1 2.11-7.746Zm1.06 10.253a6.5 6.5 0 1 0 9.108-9.275 6.5 6.5 0 0 0-9.108 9.275ZM6.03 4.97 8 6.94l1.97-1.97a.749.749 0 0 1 1.275.326.749.749 0 0 1-.215.734L9.06 8l1.97 1.97a.749.749 0 0 1-.326 1.275.749.749 0 0 1-.734-.215L8 9.06l-1.97 1.97a.749.749 0 0 1-1.275-.326.749.749 0 0 1 .215-.734L6.94 8 4.97 6.03a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018Z"/></svg>');--md-admonition-icon--danger:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M9.504.43a1.516 1.516 0 0 1 2.437 1.713L10.415 5.5h2.123c1.57 0 2.346 1.909 1.22 3.004l-7.34 7.142a1.249 1.249 0 0 1-.871.354h-.302a1.25 1.25 0 0 1-1.157-1.723L5.633 10.5H3.462c-1.57 0-2.346-1.909-1.22-3.004L9.503.429Zm1.047 1.074L3.286 8.571A.25.25 0 0 0 3.462 9H6.75a.75.75 0 0 1 .694 1.034l-1.713 4.188 6.982-6.793A.25.25 0 0 0 12.538 7H9.25a.75.75 0 0 1-.683-1.06l2.008-4.418.003-.006a.036.036 0 0 0-.004-.009l-.006-.006-.008-.001c-.003 0-.006.002-.009.004Z"/></svg>');--md-admonition-icon--bug:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M4.72.22a.75.75 0 0 1 1.06 0l1 .999a3.488 3.488 0 0 1 2.441 0l.999-1a.748.748 0 0 1 1.265.332.75.75 0 0 1-.205.729l-.775.776c.616.63.995 1.493.995 2.444v.327c0 .1-.009.197-.025.292.408.14.764.392 1.029.722l1.968-.787a.75.75 0 0 1 .556 1.392L13 7.258V9h2.25a.75.75 0 0 1 0 1.5H13v.5c0 .409-.049.806-.141 1.186l2.17.868a.75.75 0 0 1-.557 1.392l-2.184-.873A4.997 4.997 0 0 1 8 16a4.997 4.997 0 0 1-4.288-2.427l-2.183.873a.75.75 0 0 1-.558-1.392l2.17-.868A5.036 5.036 0 0 1 3 11v-.5H.75a.75.75 0 0 1 0-1.5H3V7.258L.971 6.446a.75.75 0 0 1 .558-1.392l1.967.787c.265-.33.62-.583 1.03-.722a1.677 1.677 0 0 1-.026-.292V4.5c0-.951.38-1.814.995-2.444L4.72 1.28a.75.75 0 0 1 0-1.06Zm.53 6.28a.75.75 0 0 0-.75.75V11a3.5 3.5 0 1 0 7 0V7.25a.75.75 0 0 0-.75-.75ZM6.173 5h3.654A.172.172 0 0 0 10 4.827V4.5a2 2 0 1 0-4 0v.327c0 .096.077.173.173.173Z"/></svg>');--md-admonition-icon--example:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M5 5.782V2.5h-.25a.75.75 0 0 1 0-1.5h6.5a.75.75 0 0 1 0 1.5H11v3.282l3.666 5.76C15.619 13.04 14.543 15 12.767 15H3.233c-1.776 0-2.852-1.96-1.899-3.458Zm-2.4 6.565a.75.75 0 0 0 .633 1.153h9.534a.75.75 0 0 0 .633-1.153L12.225 10.5h-8.45ZM9.5 2.5h-3V6c0 .143-.04.283-.117.403L4.73 9h6.54L9.617 6.403A.746.746 0 0 1 9.5 6Z"/></svg>');--md-admonition-icon--prerequisite:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M5 5.782V2.5h-.25a.75.75 0 0 1 0-1.5h6.5a.75.75 0 0 1 0 1.5H11v3.282l3.666 5.76C15.619 13.04 14.543 15 12.767 15H3.233c-1.776 0-2.852-1.96-1.899-3.458Zm-2.4 6.565a.75.75 0 0 0 .633 1.153h9.534a.75.75 0 0 0 .633-1.153L12.225 10.5h-8.45ZM9.5 2.5h-3V6c0 .143-.04.283-.117.403L4.73 9h6.54L9.617 6.403A.746.746 0 0 1 9.5 6Z"/></svg>');--md-admonition-icon--quote:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M1.75 2.5h10.5a.75.75 0 0 1 0 1.5H1.75a.75.75 0 0 1 0-1.5Zm4 5h8.5a.75.75 0 0 1 0 1.5h-8.5a.75.75 0 0 1 0-1.5Zm0 5h8.5a.75.75 0 0 1 0 1.5h-8.5a.75.75 0 0 1 0-1.5ZM2.5 7.75v6a.75.75 0 0 1-1.5 0v-6a.75.75 0 0 1 1.5 0Z"/></svg>');}</style>



    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="blue">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#single-cell-rna-sequencing-data-structures-in-r" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Bioinformatics Education and Services at Tufts" class="md-header__button md-logo" aria-label="Bioinformatics Education and Services at Tufts" data-md-component="logo">
      
  <img src="../../images/logo1.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Bioinformatics Education and Services at Tufts
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Formatting
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="blue"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="blue"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
</form>
      
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/best-tufts/intro_to_single_cell_rnaseq" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../00_background/" class="md-tabs__link">
          
  
  Workshop Materials

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../.." class="md-tabs__link">
          
  
  All Workshops

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Bioinformatics Education and Services at Tufts" class="md-nav__button md-logo" aria-label="Bioinformatics Education and Services at Tufts" data-md-component="logo">
      
  <img src="../../images/logo1.svg" alt="logo">

    </a>
    Bioinformatics Education and Services at Tufts
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/best-tufts/intro_to_single_cell_rnaseq" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
    
  
  
    
    
    
    
    
      
      
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_1" checked>
        
          
          <label class="md-nav__link" for="__nav_1" id="__nav_1_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Workshop Materials
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_1">
            <span class="md-nav__icon md-icon"></span>
            Workshop Materials
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../00_background/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Schedule
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../01_setup/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Set up
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Formatting
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Formatting
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#the-data" class="md-nav__link">
    The Data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#setting-up-r-environment" class="md-nav__link">
    Setting up R environment
  </a>
  
    <nav class="md-nav" aria-label="Setting up R environment">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#r-library-source" class="md-nav__link">
    R library source
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#read-in-r-packages" class="md-nav__link">
    Read in R packages
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#set-base-directory" class="md-nav__link">
    Set base directory
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#scrnaseq-count-files" class="md-nav__link">
    scRNAseq count files
  </a>
  
    <nav class="md-nav" aria-label="scRNAseq count files">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#barcodestsvgz" class="md-nav__link">
    barcodes.tsv.gz
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#featurestsvgz" class="md-nav__link">
    features.tsv.gz
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#matrixmtxgz" class="md-nav__link">
    matrix.mtx.gz
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#creating-seurat-objects" class="md-nav__link">
    Creating Seurat objects
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#seurat-object-structure" class="md-nav__link">
    Seurat object  structure
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#expression-data-assays" class="md-nav__link">
    Expression data "assays"
  </a>
  
    <nav class="md-nav" aria-label="Expression data "assays"">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#the-count-slot" class="md-nav__link">
    The count slot
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#the-data-slot" class="md-nav__link">
    The data slot
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#the-scaledata-slot" class="md-nav__link">
    The scale.data slot
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#seurat-objects-with-multiple-assays" class="md-nav__link">
    Seurat objects with multiple assays
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cell-profile-metadata" class="md-nav__link">
    Cell profile metadata
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#combining-data-sets" class="md-nav__link">
    Combining data sets
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#adding-metadata-for-downstream-analyses" class="md-nav__link">
    Adding metadata for downstream analyses
  </a>
  
    <nav class="md-nav" aria-label="Adding metadata for downstream analyses">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#novelty-score" class="md-nav__link">
    Novelty score
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#percent-mitochondrial-umi" class="md-nav__link">
    Percent mitochondrial UMI
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#finishing-touches" class="md-nav__link">
    Finishing touches
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#save-seurat-object" class="md-nav__link">
    Save seurat object
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../03_quality_control/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quality Control
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../04_integration/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Integration
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../05_clustering/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Clustering
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../06_cell_type_identification/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Cell Type Identification
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../07_differential_expression/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Differential Expression
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
    
    
    
    
    
      
      
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="">
            
  
  <span class="md-ellipsis">
    All Workshops
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            All Workshops
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    All Workshops
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#the-data" class="md-nav__link">
    The Data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#setting-up-r-environment" class="md-nav__link">
    Setting up R environment
  </a>
  
    <nav class="md-nav" aria-label="Setting up R environment">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#r-library-source" class="md-nav__link">
    R library source
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#read-in-r-packages" class="md-nav__link">
    Read in R packages
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#set-base-directory" class="md-nav__link">
    Set base directory
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#scrnaseq-count-files" class="md-nav__link">
    scRNAseq count files
  </a>
  
    <nav class="md-nav" aria-label="scRNAseq count files">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#barcodestsvgz" class="md-nav__link">
    barcodes.tsv.gz
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#featurestsvgz" class="md-nav__link">
    features.tsv.gz
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#matrixmtxgz" class="md-nav__link">
    matrix.mtx.gz
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#creating-seurat-objects" class="md-nav__link">
    Creating Seurat objects
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#seurat-object-structure" class="md-nav__link">
    Seurat object  structure
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#expression-data-assays" class="md-nav__link">
    Expression data "assays"
  </a>
  
    <nav class="md-nav" aria-label="Expression data "assays"">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#the-count-slot" class="md-nav__link">
    The count slot
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#the-data-slot" class="md-nav__link">
    The data slot
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#the-scaledata-slot" class="md-nav__link">
    The scale.data slot
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#seurat-objects-with-multiple-assays" class="md-nav__link">
    Seurat objects with multiple assays
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cell-profile-metadata" class="md-nav__link">
    Cell profile metadata
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#combining-data-sets" class="md-nav__link">
    Combining data sets
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#adding-metadata-for-downstream-analyses" class="md-nav__link">
    Adding metadata for downstream analyses
  </a>
  
    <nav class="md-nav" aria-label="Adding metadata for downstream analyses">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#novelty-score" class="md-nav__link">
    Novelty score
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#percent-mitochondrial-umi" class="md-nav__link">
    Percent mitochondrial UMI
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#finishing-touches" class="md-nav__link">
    Finishing touches
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#save-seurat-object" class="md-nav__link">
    Save seurat object
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
    <a href="https://github.com/best-tufts/intro_to_single_cell_rnaseq/blob/main/docs/intro_to_single_cell_rnaseq/02_formatting.md" title="Edit this page" class="md-content__button md-icon">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4v-2m10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1 2.1 2.1Z"/></svg>
    </a>
  
  


<h1 id="single-cell-rna-sequencing-data-structures-in-r">Single Cell RNA Sequencing Data Structures in R</h1>
<p>In this section we will describe the common data structures of singe cell RNA sequencing data (scRNAseq) data and how to read in and manipulate these data in R, specifically using the <a href="https://satijalab.org/seurat/"><code>Seurat</code></a> package. Furthermore, we will demonstrate how to read-in and merge multiple data sets. Portions of this section have been adapted from a previous Tufts HPC <a href="https://hbctraining.github.io/scRNA-seq/lessons/03_SC_quality_control-setup.html">workshop</a>.</p>
<h2 id="the-data">The Data</h2>
<p>For this workshop we will be working with a scRNAseq dataset which is part of a larger study from <a href="https://www.nature.com/articles/nbt.4042">Kang et al, 2017</a>. The data is comprised of pooled Peripheral Blood Mononuclear Cells (PBMCs) taken from lupus patients, split into control and interferon beta-treated (stimulated) conditions. The counts for this dataset is freely available from 10X Genomics and is used as part of a <a href="https://satijalab.org/seurat/archive/v3.0/immune_alignment.html">Seurat tutorial</a>.</p>
<p>Since they comprise PBMCs, we expect these data to include mononuclear immune cells, such as</p>
<ul>
<li>B cells</li>
<li>T cells</li>
<li>Natural killer cells</li>
<li>Monocytes</li>
<li>Macrophages</li>
</ul>
<p>and exclude polymorphonuclear immune cell types (neutrophils) and  nucleus lacking blood cells (erythrocytes and platelets).</p>
<h2 id="setting-up-r-environment">Setting up R environment</h2>
<h3 id="r-library-source">R library source</h3>
<p>This workshop requires a series of R libraries. In order to load these libraries we will first set their path in our R environment.</p>
<div class="highlight"><pre><span></span><code><span class="n">LIB</span><span class="o">=</span><span class="s">&#39;/cluster/tufts/hpc/tools/R/4.0.0/&#39;</span>
<span class="nf">.libPaths</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;&quot;</span><span class="p">,</span><span class="n">LIB</span><span class="p">))</span>
</code></pre></div>
<h3 id="read-in-r-packages">Read in R packages</h3>
<p>For this section, requires three R packages:</p>
<ol>
<li><code>Seurat</code> : A package for working with and analysis of scRNAseq data. Comprehensive tutorials for available analyses with the <code>Seurat</code> R package are available on the project <a href="https://satijalab.org/seurat/">website</a>.</li>
<li><code>Matrix</code>: A package for working with large data matrices.</li>
<li><code>tidyverse</code> : Actually a compendia of packages, which include functionality for "tidy" data wrangling.</li>
</ol>
<div class="highlight"><pre><span></span><code><span class="nf">library</span><span class="p">(</span><span class="n">Seurat</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">Matrix</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
</code></pre></div>
<h3 id="set-base-directory">Set base directory</h3>
<p>We will be reading in and writing files relative to our <code>intro_to_scrnaseq</code>. For simplicity, we will create an R object that is simply a character string that gives this path, and use it as a prefix for reading and writing files.</p>
<div class="highlight"><pre><span></span><code><span class="n">baseDir</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;~/intro_to_scrnaseq/&quot;</span>
</code></pre></div>
<h2 id="scrnaseq-count-files">scRNAseq count files</h2>
<p>Following pre-processing of raw files, scRNAseq count data is generally stored in directories containing three files. In this workshop will be working with two scRNAseq count data sets for each of our two condition, stored in two directories:</p>
<ol>
<li><code>ctrl_raw_feature_bc_matrix</code> : Control PBMCs</li>
<li><code>stim_raw_feature_bc_matrix</code> : Interferon beta-treated (stimulated) PBMCs</li>
</ol>
<p>Each directory contains our count data as three compressed files:</p>
<ol>
<li><code>barcodes.tsv.gz</code></li>
<li><code>features.tsv.gz</code></li>
<li><code>matrix.mtx.gz</code></li>
</ol>
<p>Next, we will read in describe each file in the <code>ctrl_raw_feature_bc_matrix</code> directory.</p>
<h3 id="barcodestsvgz"><code>barcodes.tsv.gz</code></h3>
<p>This is a text file which contains all cellular barcodes present for that sample. Barcodes are listed in the order of data presented in the matrix file (i.e. these are the column names).</p>
<div class="highlight"><pre><span></span><code><span class="n">barcodes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read.delim</span><span class="p">(</span><span class="nf">file.path</span><span class="p">(</span><span class="n">baseDir</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;data/ctrl_raw_feature_bc_matrix/barcodes.tsv.gz&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">header</span><span class="o">=</span><span class="bp">F</span><span class="p">)</span>
<span class="nf">head</span><span class="p">(</span><span class="n">barcodes</span><span class="p">)</span>
</code></pre></div>
<p><img alt="" src="../images/barcodes.png" /></p>
<h3 id="featurestsvgz"><code>features.tsv.gz</code></h3>
<p>This is a text file which contains the identifiers of the quantified genes. The source of the identifier can vary depending on what reference (i.e. Ensembl, NCBI, UCSC) you use in the quantification methods, but most often these are official gene symbols. The order of these genes corresponds to the order of the rows in the matrix file (i.e. these are the row names).</p>
<div class="highlight"><pre><span></span><code><span class="n">features</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read.delim</span><span class="p">(</span><span class="nf">file.path</span><span class="p">(</span><span class="n">baseDir</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;data/ctrl_raw_feature_bc_matrix/features.tsv.gz&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">header</span><span class="o">=</span><span class="bp">F</span><span class="p">)</span>
<span class="nf">head</span><span class="p">(</span><span class="n">features</span><span class="p">)</span>
</code></pre></div>
<p><img alt="" src="../images/features.png" /></p>
<h3 id="matrixmtxgz"><code>matrix.mtx.gz</code></h3>
<p>This is a text file which contains a matrix of count values. The rows are associated with the gene IDs above and columns correspond to the cellular barcodes. Note that there are many zero values in this matrix.</p>
<p>Note, we will read this file in R, using the <code>readMM()</code> function from the <code>Matrix</code> package. This will create a "sparse matrix" which is formatted to save space due the redundancy of 0 values. When we view this object, 0 values are shown as "."s. Below, we print the first ten rows and columns.</p>
<div class="highlight"><pre><span></span><code><span class="n">matrix</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">readMM</span><span class="p">(</span><span class="nf">file.path</span><span class="p">(</span><span class="n">baseDir</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;data/ctrl_raw_feature_bc_matrix/matrix.mtx.gz&quot;</span><span class="p">))</span>
<span class="n">matrix</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">]</span>
</code></pre></div>
<p><img alt="" src="../images/matrixObject.png" /></p>
<p>Based on the dimensions of this object, we see that there are 33,538 genes (rows), and 737,280 samples (columns) in these data.</p>
<div class="highlight"><pre><span></span><code><span class="nf">dim</span><span class="p">(</span><span class="n">matrix</span><span class="p">)</span>
</code></pre></div>
<p><img alt="" src="../images/dimMatrix.png" /></p>
<h2 id="creating-seurat-objects">Creating Seurat objects</h2>
<p>As previously mentions, in this workshop, we are going to perform the bulk of scRNAseq data manipulation and analysis using the <code>Seurat</code> R package. To this end, we are going to start by reading in the data to a <code>Seurat</code> object. A <code>Seurat</code> object has a specific data structure, for which functionality of the <code>Seurat</code> R package is built around for performing various procedures with our scRNAseq data.</p>
<p>This can be performed in two steps.</p>
<p>First, we use the <code>Read10X()</code> function from the <code>Seurat</code> package to load our scRNAseq data into R. This function assumes that our data is stored in a directory containing our three files, <code>barcodes.tsv.gz</code>, <code>features.tsv.gz</code>, <code>matrix.mtx.gz</code>, which is the standard for scRNAseq data generated by 10X Genomics. Thus, rather than reading in each file individually, we only need to set the path to the directory.</p>
<div class="highlight"><pre><span></span><code><span class="n">ctrl_counts</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">Read10X</span><span class="p">(</span><span class="n">data.dir</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">file.path</span><span class="p">(</span><span class="n">baseDir</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;data/ctrl_raw_feature_bc_matrix&quot;</span><span class="p">))</span>
</code></pre></div>
<p>Second, we use the <code>CreateSeuratObject()</code> function to convert the data as a <code>Seurat</code> object. Additionally, when we implement this function, we perform an initial data processing step by setting <code>min.features = 100</code>. This will remove cell profiles for which there are fewer than 100 genes with counts greater than 0.</p>
<div class="highlight"><pre><span></span><code><span class="n">ctrl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">CreateSeuratObject</span><span class="p">(</span><span class="n">counts</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ctrl_counts</span><span class="p">,</span>
<span class="w">                           </span><span class="n">min.features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">100</span><span class="p">)</span>
</code></pre></div>
<h2 id="seurat-object-structure">Seurat object  structure</h2>
<p>Basic features of a <code>Seurat</code> object can be viewed by simply calling the object itself.</p>
<p>For example, viewing the object output we can see that after filtering low coverage samples, we are left with 15,688 cell profiles. Which is only about 2% of the total in the raw count matrix file.</p>
<div class="highlight"><pre><span></span><code><span class="n">ctrl</span>
</code></pre></div>
<p><img alt="" src="../images/ctrlObject.png" /></p>
<h2 id="expression-data-assays">Expression data "assays"</h2>
<p>Note that as part of the output we see the term "Active assay: RNA". In <code>Seurat</code> objects, expression data is organized into these "assays", where each assay is a container of which different states of the expression data are stored separately. Every assay contains three "slots":</p>
<ul>
<li><code>RNA</code> assay<ul>
<li><code>count</code></li>
<li><code>data</code></li>
<li><code>scale.data</code></li>
</ul>
</li>
</ul>
<h3 id="the-count-slot">The <code>count</code> slot</h3>
<p>The <code>count</code> slot always contains the expression data as counts as a sparse matrix. Below are a couple of examples for accessing these data from the Seurat object.</p>
<div class="highlight"><pre><span></span><code><span class="n">countSlot</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ctrl</span><span class="o">@</span><span class="n">assays</span><span class="o">$</span><span class="n">RNA</span><span class="o">@</span><span class="n">counts</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">]</span>
<span class="n">countSlot</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">GetAssayData</span><span class="p">(</span><span class="n">ctrl</span><span class="p">,</span><span class="w"> </span><span class="n">slot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;counts&quot;</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">]</span>
<span class="n">countSlot</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">]</span>
</code></pre></div>
<p><img alt="" src="../images/countDatSlot.png" /></p>
<h3 id="the-data-slot">The <code>data</code> slot</h3>
<p>The purpose of the <code>data</code> slot is to store normalized data. We will perform normalization procedures later in this workshop, and this is commonly performed using the <code>NormalizeData()</code> <code>Seurat</code> function. However, since no normalization has been performed, the <code>data</code> slot is currently identical to the <code>count</code> slot.</p>
<div class="highlight"><pre><span></span><code><span class="n">dataSlot</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ctrl</span><span class="o">@</span><span class="n">assays</span><span class="o">$</span><span class="n">RNA</span><span class="o">@</span><span class="n">data</span>
<span class="n">dataSlot</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">GetAssayData</span><span class="p">(</span><span class="n">ctrl</span><span class="p">,</span><span class="w"> </span><span class="n">slot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;data&quot;</span><span class="p">)</span>
<span class="n">dataSlot</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">]</span>
</code></pre></div>
<p><img alt="" src="../images/countDatSlot.png" /></p>
<h3 id="the-scaledata-slot">The <code>scale.data</code> slot</h3>
<p>Numerous analytic procedures are best suited for data that has been scaled, such that the mean expression of every genes is equal to 0 with standard deviation equal to 1. This procedure is generally performed on normalized data, and the resulting "scaled" data is stored in the <code>scale.data</code> slot. Accordingly, right now this slot is an empty matrix.</p>
<div class="highlight"><pre><span></span><code><span class="n">scalSlot</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ctrl</span><span class="o">@</span><span class="n">assays</span><span class="o">$</span><span class="n">RNA</span><span class="o">@</span><span class="n">scale.data</span>
<span class="n">scalSlot</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">GetAssayData</span><span class="p">(</span><span class="n">ctrl</span><span class="p">,</span><span class="w"> </span><span class="n">slot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;scale.data&quot;</span><span class="p">)</span>
<span class="n">scalSlot</span>
</code></pre></div>
<p><img alt="" src="../images/zeroBYzero.png" /></p>
<h3 id="seurat-objects-with-multiple-assays"><code>Seurat</code> objects with multiple assays</h3>
<p>To prevent overwriting data or necessitating the creation of multiple <code>Seurat</code> objects, various procedures will create new assays for storing data. For example, later in this workshop we will perform a procedure, using the <code>SCTransform()</code> function (detailed later), which creates a new assay and the resulting assay structure of the <code>Seurat</code> object will have an assay structure:</p>
<ul>
<li>RNA assay<ul>
<li>count</li>
<li>data</li>
<li>scale.data</li>
</ul>
</li>
<li>SCT assay<ul>
<li>count</li>
<li>data</li>
<li>scale.data</li>
</ul>
</li>
</ul>
<p>We can list the available assays that have been created using the <code>Assays()</code> <code>Seurat</code> function. More importantly we can set the default assay using the <code>DefaultAssay()</code> function. This is good practice when working with <code>Seurat</code> objects containing multiple assays, to ensure that specific procedures are performed on the data for which they are intended.</p>
<div class="highlight"><pre><span></span><code><span class="nf">Assays</span><span class="p">(</span><span class="n">ctrl</span><span class="p">)</span>
<span class="nf">DefaultAssay</span><span class="p">(</span><span class="n">ctrl</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;RNA&quot;</span>
</code></pre></div>
<h2 id="cell-profile-metadata">Cell profile metadata</h2>
<p>In <code>Seurat</code> objects, the "metadata" refers to a data frame, in which we store additional information for each cell profile.</p>
<p>We can access this data from the <code>Seurat</code> object in a number of ways. For example, below we show two ways of accessing the entire metadata data frame.</p>
<div class="highlight"><pre><span></span><code><span class="n">metaData</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ctrl</span><span class="p">[[]]</span>
<span class="n">metaData</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ctrl</span><span class="o">@</span><span class="n">meta.data</span>
<span class="nf">head</span><span class="p">(</span><span class="n">metaData</span><span class="p">)</span>
</code></pre></div>
<p><img alt="" src="../images/metaData1.png" /></p>
<p>Here, the rows are in the same order as the expression matrix, which allows use to index the expression data based on this information. </p>
<p>By default, the <code>CreateSeuratObject()</code> function creates three columns in the meta.data</p>
<ul>
<li><code>orig.ident</code>: An identifier for the data set</li>
<li><code>nCount_RNA</code>: The total number of counts each cell profile</li>
<li><code>nFeature_RNA</code>: The total number of genes with counts &gt; 0. </li>
</ul>
<p>In addition to all columns there are a number of ways to extract specific columns in the metadata directly from the <code>Seurat</code> object. For example, here are two ways we can extract the <code>nCount_RNA</code> column from the metadata.</p>
<p><div class="highlight"><pre><span></span><code><span class="n">nCount</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ctrl</span><span class="o">$</span><span class="n">nCount_RNA</span>
<span class="n">nCount</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ctrl</span><span class="o">@</span><span class="n">meta.data</span><span class="o">$</span><span class="n">nCount_RNA</span>
<span class="nf">head</span><span class="p">(</span><span class="n">nCount</span><span class="p">)</span>
</code></pre></div>
<img alt="" src="../images/nCount.png" /></p>
<p>Alternatively, there are many ways to extract multiple metadata columns from the <code>Seurat</code> object.</p>
<p><div class="highlight"><pre><span></span><code><span class="n">multCols</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ctrl</span><span class="o">@</span><span class="n">meta.data</span><span class="p">[,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;nCount_RNA&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;nFeature_RNA&quot;</span><span class="p">)]</span>
<span class="n">multCols</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ctrl</span><span class="p">[[</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;nCount_RNA&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;nFeature_RNA&quot;</span><span class="p">)]]</span>
<span class="nf">head</span><span class="p">(</span><span class="n">multCols</span><span class="p">)</span>
</code></pre></div>
<img alt="" src="../images/metaData2.png" /></p>
<h2 id="combining-data-sets">Combining data sets</h2>
<p>It is common for scRNAseq projects to include multiple samples profiled separately. In our case we have two separate data sets, comprising profiles from "control" and "stimulated" PBMC samples. Accordingly, we're going to combine these data into a single <code>Seurat</code> object, while assigning information as to their source in the metadata.</p>
<p>Here, we use a <code>for</code> loop to read in each of the two data sets. Within each we assign an identity to each data set using the <code>project</code> argument.</p>
<div class="highlight"><pre><span></span><code><span class="nf">for </span><span class="p">(</span><span class="n">file</span><span class="w"> </span><span class="n">in</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;ctrl_raw_feature_bc_matrix&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;stim_raw_feature_bc_matrix&quot;</span><span class="p">)){</span>
<span class="w">  </span><span class="n">seurat_data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">Read10X</span><span class="p">(</span><span class="n">data.dir</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">file.path</span><span class="p">(</span><span class="n">baseDir</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;data&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">file</span><span class="p">))</span>
<span class="w">  </span><span class="n">seurat_obj</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">CreateSeuratObject</span><span class="p">(</span><span class="n">counts</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">seurat_data</span><span class="p">,</span><span class="w"> </span>
<span class="w">                                   </span><span class="n">min.features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">100</span><span class="p">,</span><span class="w"> </span>
<span class="w">                                   </span><span class="n">project</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">file</span><span class="p">)</span>
<span class="w">  </span><span class="nf">assign</span><span class="p">(</span><span class="n">file</span><span class="p">,</span><span class="w"> </span><span class="n">seurat_obj</span><span class="p">)</span>
<span class="p">}</span>
</code></pre></div>
<p>We have just created two <code>Seurat</code> objects, <code>ctrl_raw_feature_bc_matrix</code> and <code>stim_raw_feature_bc_matrix</code>. And if we check the metadata of <code>ctrl_raw_feature_bc_matrix</code> to see that its source has been saved in the <code>orig.identity</code> column.</p>
<div class="highlight"><pre><span></span><code><span class="nf">head</span><span class="p">(</span><span class="n">ctrl_raw_feature_bc_matrix</span><span class="o">@</span><span class="n">meta.data</span><span class="p">)</span>
</code></pre></div>
<p><img alt="" src="../images/metaData3.png" /></p>
<p>Finally, we merge the two <code>Seurat</code> objects with the <code>merge()</code> function below. Since each data set will contain likely some of the same cell profile identifiers, e.g. "AAACATACAATGCC-1", we add a prefix to these identifiers using the <code>add.cell.id</code> argument.</p>
<div class="highlight"><pre><span></span><code><span class="n">merged_seurat</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">merge</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ctrl_raw_feature_bc_matrix</span><span class="p">,</span><span class="w"> </span>
<span class="w">                       </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">stim_raw_feature_bc_matrix</span><span class="p">,</span><span class="w"> </span>
<span class="w">                       </span><span class="n">add.cell.id</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;ctrl&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;stim&quot;</span><span class="p">))</span>
</code></pre></div>
<p>Using the <code>orig.identity</code> column we can check how many cell profiles from each data set are contained in the merged data set.</p>
<div class="highlight"><pre><span></span><code><span class="nf">table</span><span class="p">(</span><span class="n">merged_seurat</span><span class="o">@</span><span class="n">meta.data</span><span class="o">$</span><span class="n">orig.ident</span><span class="p">)</span>
</code></pre></div>
<p><img alt="" src="../images/mergeCounts.png" /></p>
<h2 id="adding-metadata-for-downstream-analyses">Adding metadata for downstream analyses</h2>
<p>Now that we have the fully merged scRNAseq data, we will add a couple additional columns to our metadata, which we will later use for quality control. These columns comprise calculations of two metrics, which are indicative the quality of individual cell profiles.</p>
<ul>
<li><strong>Novelty score</strong>: this metric with give us an idea of the complexity of our dataset (more genes detected per UMI, more complex our data).</li>
<li><strong>Percent mitochondrial UMI</strong>: this metric will give us a percentage of cell reads originating from the mitochondrial genes</li>
</ul>
<h3 id="novelty-score">Novelty score</h3>
<p>This value is quite easy to calculate, as we take the <strong>log10</strong> of the number of genes detected per cell and the <strong>log10</strong> of the number of UMIs per cell, then divide the <strong>log10</strong> number of genes by the log10 number of UMIs. The novelty score and how it relates to complexity of the RNA species, is described in more detail later in this lesson.</p>
<div class="highlight"><pre><span></span><code><span class="n">merged_seurat</span><span class="o">$</span><span class="n">log10GenesPerUMI</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">log10</span><span class="p">(</span><span class="n">merged_seurat</span><span class="o">$</span><span class="n">nFeature_RNA</span><span class="p">)</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="nf">log10</span><span class="p">(</span><span class="n">merged_seurat</span><span class="o">$</span><span class="n">nCount_RNA</span><span class="p">)</span>
</code></pre></div>
<h3 id="percent-mitochondrial-umi">Percent mitochondrial UMI</h3>
<p>Seurat has a convenient function that allows us to calculate the **proportion of transcripts mapping to mitochondrial genes(. The <code>PercentageFeatureSet()</code> function takes in a pattern argument and searches through all gene identifiers in the dataset for that pattern. Since we are looking for mitochondrial genes, we are searching any gene identifiers that begin with the pattern "MT-". For each cell, the function takes the sum of counts across all genes (features) belonging to the "Mt-" set, and then divides by the count sum for all genes (features). This value is multiplied by 100 to obtain a percentage value.</p>
<div class="highlight"><pre><span></span><code><span class="n">merged_seurat</span><span class="o">$</span><span class="n">percMitoUMI</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">PercentageFeatureSet</span><span class="p">(</span><span class="n">object</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">merged_seurat</span><span class="p">,</span><span class="w"> </span><span class="n">pattern</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;^MT-&quot;</span><span class="p">)</span>
</code></pre></div>
<h2 id="finishing-touches">Finishing touches</h2>
<p>Finally, we will make a few additional changes to our metadata to facilitate downstream analyses.</p>
<p>First, we will create a new metadata column with simplified identifers for each of the merged data sets. We can do this easily, by taking the first 4 characters of the values from the <code>orig.identity</code> column, and assigning them to a new column.</p>
<div class="highlight"><pre><span></span><code><span class="n">merged_seurat</span><span class="o">$</span><span class="n">sample</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">substr</span><span class="p">(</span><span class="n">merged_seurat</span><span class="o">@</span><span class="n">meta.data</span><span class="o">$</span><span class="n">orig.ident</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">)</span>

<span class="nf">table</span><span class="p">(</span><span class="n">merged_seurat</span><span class="o">$</span><span class="n">sample</span><span class="p">)</span>
</code></pre></div>
<p><img alt="" src="../images/sampTab.png" /></p>
<p>Next, we will rename several of metadata column names short, so they are shorter and more intuitive (to us at least).</p>
<div class="highlight"><pre><span></span><code><span class="n">merged_seurat</span><span class="o">@</span><span class="n">meta.data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">merged_seurat</span><span class="o">@</span><span class="n">meta.data</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="n">dplyr</span><span class="o">::</span><span class="nf">rename</span><span class="p">(</span><span class="n">seq_folder</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">orig.ident</span><span class="p">,</span>
<span class="w">                </span><span class="n">nUMI</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">nCount_RNA</span><span class="p">,</span>
<span class="w">                </span><span class="n">nGene</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">nFeature_RNA</span><span class="p">)</span>
</code></pre></div>
<p>Lastly we can check the final formatted metadata dataframe with all of our newly added and formatted metadata columns. These are a lot of columns to view in our console and html output, so we'll just use the <code>View()</code> R function.</p>
<div class="highlight"><pre><span></span><code>View(merged_seurat@meta.data)
</code></pre></div>
<p><img alt="" src="../images/metaData4.png" /></p>
<h2 id="save-seurat-object">Save seurat object</h2>
<p>That's it! At this point, we can save our merged data set and move on to quality control.</p>
<div class="highlight"><pre><span></span><code><span class="nf">saveRDS</span><span class="p">(</span><span class="n">merged_seurat</span><span class="p">,</span><span class="w"> </span><span class="nf">file.path</span><span class="p">(</span><span class="n">baseDir</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;results/merged_seurat.rds&quot;</span><span class="p">))</span>
</code></pre></div>
<p>You can now open "03_quality_control.Rmd" to continue on the the next section.</p>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../01_setup/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Set up">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                Set up
              </div>
            </div>
          </a>
        
        
          
          <a href="../03_quality_control/" class="md-footer__link md-footer__link--next" aria-label="Next: Quality Control">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                Quality Control
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tabs", "navigation.tabs.sticky", "navigation.sections", "content.action.edit", "navigation.footer", "content.code.copy"], "search": "../../assets/javascripts/workers/search.f886a092.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.aecac24b.min.js"></script>
      
        <script src="../../javascripts/mathjax.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
        <script src="https://unpkg.com/tablesort@5.3.0/dist/tablesort.min.js"></script>
      
        <script src="../../javascripts/tablesort.js"></script>
      
    
  </body>
</html>