
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../03_quality_control/">
      
      
        <link rel="next" href="../05_clustering/">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.4.7">
    
    
      
        <title>Integration - Bioinformatics Education and Services at Tufts</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.4b4a2bd9.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.356b1318.min.css">
      
      
  
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
  
  <style>:root{--md-admonition-icon--note:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M1 7.775V2.75C1 1.784 1.784 1 2.75 1h5.025c.464 0 .91.184 1.238.513l6.25 6.25a1.75 1.75 0 0 1 0 2.474l-5.026 5.026a1.75 1.75 0 0 1-2.474 0l-6.25-6.25A1.752 1.752 0 0 1 1 7.775Zm1.5 0c0 .066.026.13.073.177l6.25 6.25a.25.25 0 0 0 .354 0l5.025-5.025a.25.25 0 0 0 0-.354l-6.25-6.25a.25.25 0 0 0-.177-.073H2.75a.25.25 0 0 0-.25.25ZM6 5a1 1 0 1 1 0 2 1 1 0 0 1 0-2Z"/></svg>');--md-admonition-icon--abstract:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M2.5 1.75v11.5c0 .138.112.25.25.25h3.17a.75.75 0 0 1 0 1.5H2.75A1.75 1.75 0 0 1 1 13.25V1.75C1 .784 1.784 0 2.75 0h8.5C12.216 0 13 .784 13 1.75v7.736a.75.75 0 0 1-1.5 0V1.75a.25.25 0 0 0-.25-.25h-8.5a.25.25 0 0 0-.25.25Zm13.274 9.537v-.001l-4.557 4.45a.75.75 0 0 1-1.055-.008l-1.943-1.95a.75.75 0 0 1 1.062-1.058l1.419 1.425 4.026-3.932a.75.75 0 1 1 1.048 1.074ZM4.75 4h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5ZM4 7.75A.75.75 0 0 1 4.75 7h2a.75.75 0 0 1 0 1.5h-2A.75.75 0 0 1 4 7.75Z"/></svg>');--md-admonition-icon--info:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M0 8a8 8 0 1 1 16 0A8 8 0 0 1 0 8Zm8-6.5a6.5 6.5 0 1 0 0 13 6.5 6.5 0 0 0 0-13ZM6.5 7.75A.75.75 0 0 1 7.25 7h1a.75.75 0 0 1 .75.75v2.75h.25a.75.75 0 0 1 0 1.5h-2a.75.75 0 0 1 0-1.5h.25v-2h-.25a.75.75 0 0 1-.75-.75ZM8 6a1 1 0 1 1 0-2 1 1 0 0 1 0 2Z"/></svg>');--md-admonition-icon--tip:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M3.499.75a.75.75 0 0 1 1.5 0v.996C5.9 2.903 6.793 3.65 7.662 4.376l.24.202c-.036-.694.055-1.422.426-2.163C9.1.873 10.794-.045 12.622.26 14.408.558 16 1.94 16 4.25c0 1.278-.954 2.575-2.44 2.734l.146.508.065.22c.203.701.412 1.455.476 2.226.142 1.707-.4 3.03-1.487 3.898C11.714 14.671 10.27 15 8.75 15h-6a.75.75 0 0 1 0-1.5h1.376a4.484 4.484 0 0 1-.563-1.191 3.835 3.835 0 0 1-.05-2.063 4.647 4.647 0 0 1-2.025-.293.75.75 0 0 1 .525-1.406c1.357.507 2.376-.006 2.698-.318l.009-.01a.747.747 0 0 1 1.06 0 .748.748 0 0 1-.012 1.074c-.912.92-.992 1.835-.768 2.586.221.74.745 1.337 1.196 1.621H8.75c1.343 0 2.398-.296 3.074-.836.635-.507 1.036-1.31.928-2.602-.05-.603-.216-1.224-.422-1.93l-.064-.221c-.12-.407-.246-.84-.353-1.29a2.425 2.425 0 0 1-.507-.441 3.075 3.075 0 0 1-.633-1.248.75.75 0 0 1 1.455-.364c.046.185.144.436.31.627.146.168.353.305.712.305.738 0 1.25-.615 1.25-1.25 0-1.47-.95-2.315-2.123-2.51-1.172-.196-2.227.387-2.706 1.345-.46.92-.27 1.774.019 3.062l.042.19a.884.884 0 0 1 .01.05c.348.443.666.949.94 1.553a.75.75 0 1 1-1.365.62c-.553-1.217-1.32-1.94-2.3-2.768L6.7 5.527c-.814-.68-1.75-1.462-2.692-2.619a3.737 3.737 0 0 0-1.023.88c-.406.495-.663 1.036-.722 1.508.116.122.306.21.591.239.388.038.797-.06 1.032-.19a.75.75 0 0 1 .728 1.31c-.515.287-1.23.439-1.906.373-.682-.067-1.473-.38-1.879-1.193L.75 5.677V5.5c0-.984.48-1.94 1.077-2.664.46-.559 1.05-1.055 1.673-1.353V.75Z"/></svg>');--md-admonition-icon--success:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z"/></svg>');--md-admonition-icon--question:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M0 8a8 8 0 1 1 16 0A8 8 0 0 1 0 8Zm8-6.5a6.5 6.5 0 1 0 0 13 6.5 6.5 0 0 0 0-13ZM6.92 6.085h.001a.749.749 0 1 1-1.342-.67c.169-.339.436-.701.849-.977C6.845 4.16 7.369 4 8 4a2.756 2.756 0 0 1 1.637.525c.503.377.863.965.863 1.725 0 .448-.115.83-.329 1.15-.205.307-.47.513-.692.662-.109.072-.22.138-.313.195l-.006.004a6.24 6.24 0 0 0-.26.16.952.952 0 0 0-.276.245.75.75 0 0 1-1.248-.832c.184-.264.42-.489.692-.661.103-.067.207-.132.313-.195l.007-.004c.1-.061.182-.11.258-.161a.969.969 0 0 0 .277-.245C8.96 6.514 9 6.427 9 6.25a.612.612 0 0 0-.262-.525A1.27 1.27 0 0 0 8 5.5c-.369 0-.595.09-.74.187a1.01 1.01 0 0 0-.34.398ZM9 11a1 1 0 1 1-2 0 1 1 0 0 1 2 0Z"/></svg>');--md-admonition-icon--warning:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M6.457 1.047c.659-1.234 2.427-1.234 3.086 0l6.082 11.378A1.75 1.75 0 0 1 14.082 15H1.918a1.75 1.75 0 0 1-1.543-2.575Zm1.763.707a.25.25 0 0 0-.44 0L1.698 13.132a.25.25 0 0 0 .22.368h12.164a.25.25 0 0 0 .22-.368Zm.53 3.996v2.5a.75.75 0 0 1-1.5 0v-2.5a.75.75 0 0 1 1.5 0ZM9 11a1 1 0 1 1-2 0 1 1 0 0 1 2 0Z"/></svg>');--md-admonition-icon--failure:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M2.344 2.343h-.001a8 8 0 0 1 11.314 11.314A8.002 8.002 0 0 1 .234 10.089a8 8 0 0 1 2.11-7.746Zm1.06 10.253a6.5 6.5 0 1 0 9.108-9.275 6.5 6.5 0 0 0-9.108 9.275ZM6.03 4.97 8 6.94l1.97-1.97a.749.749 0 0 1 1.275.326.749.749 0 0 1-.215.734L9.06 8l1.97 1.97a.749.749 0 0 1-.326 1.275.749.749 0 0 1-.734-.215L8 9.06l-1.97 1.97a.749.749 0 0 1-1.275-.326.749.749 0 0 1 .215-.734L6.94 8 4.97 6.03a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018Z"/></svg>');--md-admonition-icon--danger:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M9.504.43a1.516 1.516 0 0 1 2.437 1.713L10.415 5.5h2.123c1.57 0 2.346 1.909 1.22 3.004l-7.34 7.142a1.249 1.249 0 0 1-.871.354h-.302a1.25 1.25 0 0 1-1.157-1.723L5.633 10.5H3.462c-1.57 0-2.346-1.909-1.22-3.004L9.503.429Zm1.047 1.074L3.286 8.571A.25.25 0 0 0 3.462 9H6.75a.75.75 0 0 1 .694 1.034l-1.713 4.188 6.982-6.793A.25.25 0 0 0 12.538 7H9.25a.75.75 0 0 1-.683-1.06l2.008-4.418.003-.006a.036.036 0 0 0-.004-.009l-.006-.006-.008-.001c-.003 0-.006.002-.009.004Z"/></svg>');--md-admonition-icon--bug:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M4.72.22a.75.75 0 0 1 1.06 0l1 .999a3.488 3.488 0 0 1 2.441 0l.999-1a.748.748 0 0 1 1.265.332.75.75 0 0 1-.205.729l-.775.776c.616.63.995 1.493.995 2.444v.327c0 .1-.009.197-.025.292.408.14.764.392 1.029.722l1.968-.787a.75.75 0 0 1 .556 1.392L13 7.258V9h2.25a.75.75 0 0 1 0 1.5H13v.5c0 .409-.049.806-.141 1.186l2.17.868a.75.75 0 0 1-.557 1.392l-2.184-.873A4.997 4.997 0 0 1 8 16a4.997 4.997 0 0 1-4.288-2.427l-2.183.873a.75.75 0 0 1-.558-1.392l2.17-.868A5.036 5.036 0 0 1 3 11v-.5H.75a.75.75 0 0 1 0-1.5H3V7.258L.971 6.446a.75.75 0 0 1 .558-1.392l1.967.787c.265-.33.62-.583 1.03-.722a1.677 1.677 0 0 1-.026-.292V4.5c0-.951.38-1.814.995-2.444L4.72 1.28a.75.75 0 0 1 0-1.06Zm.53 6.28a.75.75 0 0 0-.75.75V11a3.5 3.5 0 1 0 7 0V7.25a.75.75 0 0 0-.75-.75ZM6.173 5h3.654A.172.172 0 0 0 10 4.827V4.5a2 2 0 1 0-4 0v.327c0 .096.077.173.173.173Z"/></svg>');--md-admonition-icon--example:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M5 5.782V2.5h-.25a.75.75 0 0 1 0-1.5h6.5a.75.75 0 0 1 0 1.5H11v3.282l3.666 5.76C15.619 13.04 14.543 15 12.767 15H3.233c-1.776 0-2.852-1.96-1.899-3.458Zm-2.4 6.565a.75.75 0 0 0 .633 1.153h9.534a.75.75 0 0 0 .633-1.153L12.225 10.5h-8.45ZM9.5 2.5h-3V6c0 .143-.04.283-.117.403L4.73 9h6.54L9.617 6.403A.746.746 0 0 1 9.5 6Z"/></svg>');--md-admonition-icon--prerequisite:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M5 5.782V2.5h-.25a.75.75 0 0 1 0-1.5h6.5a.75.75 0 0 1 0 1.5H11v3.282l3.666 5.76C15.619 13.04 14.543 15 12.767 15H3.233c-1.776 0-2.852-1.96-1.899-3.458Zm-2.4 6.565a.75.75 0 0 0 .633 1.153h9.534a.75.75 0 0 0 .633-1.153L12.225 10.5h-8.45ZM9.5 2.5h-3V6c0 .143-.04.283-.117.403L4.73 9h6.54L9.617 6.403A.746.746 0 0 1 9.5 6Z"/></svg>');--md-admonition-icon--quote:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M1.75 2.5h10.5a.75.75 0 0 1 0 1.5H1.75a.75.75 0 0 1 0-1.5Zm4 5h8.5a.75.75 0 0 1 0 1.5h-8.5a.75.75 0 0 1 0-1.5Zm0 5h8.5a.75.75 0 0 1 0 1.5h-8.5a.75.75 0 0 1 0-1.5ZM2.5 7.75v6a.75.75 0 0 1-1.5 0v-6a.75.75 0 0 1 1.5 0Z"/></svg>');}</style>



    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="blue">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#single-cell-rna-sequencing-normalization-dimensionality-reduction-and-integration" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Bioinformatics Education and Services at Tufts" class="md-header__button md-logo" aria-label="Bioinformatics Education and Services at Tufts" data-md-component="logo">
      
  <img src="../../images/logo1.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Bioinformatics Education and Services at Tufts
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Integration
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="blue"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="blue"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
</form>
      
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/best-tufts/intro_to_single_cell_rnaseq" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../00_background/" class="md-tabs__link">
          
  
  Workshop Materials

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../.." class="md-tabs__link">
          
  
  All Workshops

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Bioinformatics Education and Services at Tufts" class="md-nav__button md-logo" aria-label="Bioinformatics Education and Services at Tufts" data-md-component="logo">
      
  <img src="../../images/logo1.svg" alt="logo">

    </a>
    Bioinformatics Education and Services at Tufts
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/best-tufts/intro_to_single_cell_rnaseq" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
    
  
  
    
    
    
    
    
      
      
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_1" checked>
        
          
          <label class="md-nav__link" for="__nav_1" id="__nav_1_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Workshop Materials
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_1">
            <span class="md-nav__icon md-icon"></span>
            Workshop Materials
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../00_background/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Schedule
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../01_setup/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Set up
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../02_formatting/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Formatting
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../03_quality_control/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quality Control
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Integration
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Integration
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#setting-up-r-environment" class="md-nav__link">
    Setting up R environment
  </a>
  
    <nav class="md-nav" aria-label="Setting up R environment">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#r-library-source" class="md-nav__link">
    R library source
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#read-in-r-packages" class="md-nav__link">
    Read in R packages
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#set-base-directory" class="md-nav__link">
    Set base directory
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#read-in-seurat-object" class="md-nav__link">
    Read in Seurat object
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#expression-normalization" class="md-nav__link">
    Expression normalization
  </a>
  
    <nav class="md-nav" aria-label="Expression normalization">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#log-transformation" class="md-nav__link">
    Log-transformation
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cleaning-up-r-environment" class="md-nav__link">
    Cleaning up R environment
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dimensionality-reduction" class="md-nav__link">
    Dimensionality reduction
  </a>
  
    <nav class="md-nav" aria-label="Dimensionality reduction">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#feature-selection" class="md-nav__link">
    Feature selection
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#expression-scaling" class="md-nav__link">
    Expression scaling
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#uniform-manifold-approximation" class="md-nav__link">
    Uniform manifold approximation
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#alternative-expression-normalization-and-scaling-with-sctransform" class="md-nav__link">
    Alternative expression normalization and scaling with SCTransform()
  </a>
  
    <nav class="md-nav" aria-label="Alternative expression normalization and scaling with SCTransform()">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#read-in-sctransorm-seurat-object" class="md-nav__link">
    Read in SCTransorm() Seurat object
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#data-integration" class="md-nav__link">
    Data Integration
  </a>
  
    <nav class="md-nav" aria-label="Data Integration">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#canonical-correlation-analysis-cca" class="md-nav__link">
    Canonical Correlation Analysis (CCA)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mutual-nearest-neighbors-mnn" class="md-nav__link">
    Mutual Nearest Neighbors (MNN)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#select-integration-features" class="md-nav__link">
    Select integration features
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#prepare-the-sct-list-object-for-integration" class="md-nav__link">
    Prepare the SCT list object for integration
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#perform-cca-and-identify-mutual-nearest-neighbots" class="md-nav__link">
    Perform CCA and identify mutual nearest neighbots
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#integrate-data" class="md-nav__link">
    Integrate data
  </a>
  
    <nav class="md-nav" aria-label="Integrate data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#read-in-integrated-seurat-object" class="md-nav__link">
    Read in integrated Seurat object
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#check-integrated-data-set" class="md-nav__link">
    Check integrated data set
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../05_clustering/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Clustering
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../06_cell_type_identification/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Cell Type Identification
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../07_differential_expression/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Differential Expression
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
    
    
    
    
    
      
      
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="">
            
  
  <span class="md-ellipsis">
    All Workshops
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            All Workshops
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    All Workshops
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#setting-up-r-environment" class="md-nav__link">
    Setting up R environment
  </a>
  
    <nav class="md-nav" aria-label="Setting up R environment">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#r-library-source" class="md-nav__link">
    R library source
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#read-in-r-packages" class="md-nav__link">
    Read in R packages
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#set-base-directory" class="md-nav__link">
    Set base directory
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#read-in-seurat-object" class="md-nav__link">
    Read in Seurat object
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#expression-normalization" class="md-nav__link">
    Expression normalization
  </a>
  
    <nav class="md-nav" aria-label="Expression normalization">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#log-transformation" class="md-nav__link">
    Log-transformation
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cleaning-up-r-environment" class="md-nav__link">
    Cleaning up R environment
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dimensionality-reduction" class="md-nav__link">
    Dimensionality reduction
  </a>
  
    <nav class="md-nav" aria-label="Dimensionality reduction">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#feature-selection" class="md-nav__link">
    Feature selection
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#expression-scaling" class="md-nav__link">
    Expression scaling
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#uniform-manifold-approximation" class="md-nav__link">
    Uniform manifold approximation
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#alternative-expression-normalization-and-scaling-with-sctransform" class="md-nav__link">
    Alternative expression normalization and scaling with SCTransform()
  </a>
  
    <nav class="md-nav" aria-label="Alternative expression normalization and scaling with SCTransform()">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#read-in-sctransorm-seurat-object" class="md-nav__link">
    Read in SCTransorm() Seurat object
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#data-integration" class="md-nav__link">
    Data Integration
  </a>
  
    <nav class="md-nav" aria-label="Data Integration">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#canonical-correlation-analysis-cca" class="md-nav__link">
    Canonical Correlation Analysis (CCA)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mutual-nearest-neighbors-mnn" class="md-nav__link">
    Mutual Nearest Neighbors (MNN)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#select-integration-features" class="md-nav__link">
    Select integration features
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#prepare-the-sct-list-object-for-integration" class="md-nav__link">
    Prepare the SCT list object for integration
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#perform-cca-and-identify-mutual-nearest-neighbots" class="md-nav__link">
    Perform CCA and identify mutual nearest neighbots
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#integrate-data" class="md-nav__link">
    Integrate data
  </a>
  
    <nav class="md-nav" aria-label="Integrate data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#read-in-integrated-seurat-object" class="md-nav__link">
    Read in integrated Seurat object
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#check-integrated-data-set" class="md-nav__link">
    Check integrated data set
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
    <a href="https://github.com/best-tufts/intro_to_single_cell_rnaseq/blob/main/docs/intro_to_single_cell_rnaseq/04_integration.md" title="Edit this page" class="md-content__button md-icon">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4v-2m10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1 2.1 2.1Z"/></svg>
    </a>
  
  


<h1 id="single-cell-rna-sequencing-normalization-dimensionality-reduction-and-integration">Single Cell RNA Sequencing Normalization, Dimensionality Reduction, and Integration</h1>
<p>In this section we will describe procedures for processing scRNAseq data to facilitate downstream analysis, particularly clustering of scRNAseq profiles to identify cell populations. These procedures include</p>
<ul>
<li>Normalization</li>
<li>Dimensionality Reduction</li>
<li>Integration</li>
</ul>
<p>We will perform these procedures on our two-sample combined, formatted, and QC'd PBMC scRNAseq data set [<code>Seurat</code>] object generated in the previous sections. Portions of this section have been adapted from  previous Tufts HPC describing <a href="https://hbctraining.github.io/scRNA-seq_online/lessons/06_SC_SCT_normalization.html">normalization</a> and <a href="https://hbctraining.github.io/scRNA-seq_online/lessons/06_integration.html">integration</a>.</p>
<h2 id="setting-up-r-environment">Setting up R environment</h2>
<p>We begin by setting up our R environment similar to the previous sections.</p>
<h3 id="r-library-source">R library source</h3>
<p>We will be reading in and writing files relative to our <code>intro_to_scrnaseq</code>. For simplicity, we will create an R object that is simply a character string that gives this path, and use it as a prefix for reading and writing files.</p>
<div class="highlight"><pre><span></span><code><span class="n">LIB</span><span class="o">=</span><span class="s">&#39;/cluster/tufts/hpc/tools/R/4.0.0/&#39;</span>
<span class="nf">.libPaths</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;&quot;</span><span class="p">,</span><span class="n">LIB</span><span class="p">))</span>
</code></pre></div>
<h3 id="read-in-r-packages">Read in R packages</h3>
<p>For this section, requires three R packages:</p>
<ol>
<li><code>Seurat</code> : A package for working with and analysis of scRNAseq data. Comprehensive tutorials for available analyses with the <code>Seurat</code> R package are available on the project <a href="https://satijalab.org/seurat/">website</a>.</li>
<li><code>ggplot2</code>: The standard package for creating plots in R. Much of the plotting functions are wrappers for <code>ggplot2</code> functionality.</li>
<li><code>cowplot</code> : A nice package for combining plots into a single figure. Specifically we will make use of of the <code>plot_grid()</code> function.</li>
</ol>
<div class="highlight"><pre><span></span><code><span class="nf">library</span><span class="p">(</span><span class="n">Seurat</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">cowplot</span><span class="p">)</span>
</code></pre></div>
<h3 id="set-base-directory">Set base directory</h3>
<div class="highlight"><pre><span></span><code><span class="n">baseDir</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;~/intro_to_scrnaseq/&quot;</span>
</code></pre></div>
<h2 id="read-in-seurat-object">Read in <code>Seurat</code> object</h2>
<div class="highlight"><pre><span></span><code><span class="n">filtered_seurat</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">readRDS</span><span class="p">(</span><span class="nf">file.path</span><span class="p">(</span><span class="n">baseDir</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;results/merged_filtered_seurat.rds&quot;</span><span class="p">))</span>
</code></pre></div>
<h2 id="expression-normalization">Expression normalization</h2>
<p>Even for the same types of cells, the total number of UMIs that are counted is highly variable cell profile to cell profile. Accordingly, the UMI counts for each gene are not immediately comparable as they are proportional to the total number of reads. When performing data normalization we account for this variation by a dividing the reads from each cell by a scaling factor that is a function of the total number of reads of that cell.</p>
<h3 id="log-transformation">Log-transformation</h3>
<p>The variance of normalized counts are affected by their magnitude. By performing log-transformation, we partially account for this source of variance, making gene profiles more immediately comparable. </p>
<p>Data normalization and log-transformation can be performed in a single step with the <code>Seurat</code> package with the <code>NormalizeData()</code> function. Below we will run this function and create a new <code>Seurat</code> object, <code>norm_seurat</code>.</p>
<div class="highlight"><pre><span></span><code><span class="n">norm_seurat</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">NormalizeData</span><span class="p">(</span><span class="n">filtered_seurat</span><span class="p">)</span>
</code></pre></div>
<p>Following <code>NormalizeData()</code>, the <code>data</code> "slot" of our <code>RNA</code> assay is not populated by our normalized and log-transformed count data. We can visualize this using the <code>VlnPlot()</code> <code>Seurat</code> function, where we specify the source of gene expression data using the <code>slot</code> argument. Below we visualize CD63 expression.</p>
<div class="highlight"><pre><span></span><code><span class="c1">## Plot CD63 counts</span>
<span class="n">gPlotCount</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">VlnPlot</span><span class="p">(</span><span class="n">object</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">norm_seurat</span><span class="p">,</span><span class="w"> </span>
<span class="w">                      </span><span class="n">features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;CD63&quot;</span><span class="p">,</span>
<span class="w">                      </span><span class="n">group.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;sample&quot;</span><span class="p">,</span>
<span class="w">                      </span><span class="n">assay</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;RNA&quot;</span><span class="p">,</span>
<span class="w">                      </span><span class="n">slot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;count&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">ggtitle</span><span class="p">(</span><span class="s">&quot;CD63 Counts&quot;</span><span class="p">,</span>
<span class="w">          </span><span class="n">subtitle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;RNA Assay&quot;</span><span class="p">)</span>

<span class="c1">## Plot CD63 normalized Expression</span>
<span class="n">gPlotNorm</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">VlnPlot</span><span class="p">(</span><span class="n">object</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">norm_seurat</span><span class="p">,</span><span class="w"> </span>
<span class="w">                      </span><span class="n">features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;CD63&quot;</span><span class="p">,</span>
<span class="w">                      </span><span class="n">group.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;sample&quot;</span><span class="p">,</span>
<span class="w">                      </span><span class="n">assay</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;RNA&quot;</span><span class="p">,</span>
<span class="w">                      </span><span class="n">slot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;data&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">ggtitle</span><span class="p">(</span><span class="s">&quot;CD63 Norm. Expression&quot;</span><span class="p">,</span>
<span class="w">          </span><span class="n">subtitle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;RNA Assay&quot;</span><span class="p">)</span>

<span class="c1">## Combine plots</span>
<span class="nf">plot_grid</span><span class="p">(</span><span class="n">gPlotCount</span><span class="p">,</span><span class="w"> </span><span class="n">gPlotNorm</span><span class="p">,</span><span class="w"> </span><span class="n">nrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">)</span>
</code></pre></div>
<p><img alt="" src="../images/NormExp1.png" /></p>
<h3 id="cleaning-up-r-environment">Cleaning up R environment</h3>
<p><code>Seurat</code> objects contain a lot of data and become quite memory expensive. Especially when working on machines with limited available memory, removing objects we are done with can save headaches down the line. We are done with <code>filtered_seurat</code> so we can now go ahead and remove it.</p>
<div class="highlight"><pre><span></span><code><span class="nf">rm</span><span class="p">(</span><span class="n">filtered_seurat</span><span class="p">)</span>
<span class="nf">gc</span><span class="p">()</span>
</code></pre></div>
<h2 id="dimensionality-reduction">Dimensionality reduction</h2>
<p>The purpose of dimensionality reduction to capture the majority of variability in our data with fewer features than the 1,000s of genes in our data. By doing so we emphasize sources of variability that are indicative of similar and diverging cell types in our data. We can then use these new variables as input in numerous machine learning algorithms, such many of those employed for cluster analysis. Specifically, through Principal Component Analysis (PCA), we can reduce our matrix of 1,000's to 10's of features (i.e. principal components) that capture the majority of variabilty in our data.</p>
<h3 id="feature-selection">Feature selection</h3>
<p>Our <code>Seurat</code> objects is currently comprised of 14,065 features. Prior to running PCA, we will first identify 2,000 genes with high variance in our data. The justification of this is that highly variable genes should capture differences in expression across cell populations. Moreover, is makes calculation of our principal components less computationally expensive. We can perform feature selection using the <code>FindVariableFeatures()</code> <code>Seurat</code> function.</p>
<div class="highlight"><pre><span></span><code><span class="n">norm_seurat</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">FindVariableFeatures</span><span class="p">(</span><span class="n">norm_seurat</span><span class="p">,</span><span class="w"> </span>
<span class="w">                                     </span><span class="n">selection.method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;vst&quot;</span><span class="p">,</span>
<span class="w">                                     </span><span class="n">nfeatures</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3000</span><span class="p">,</span><span class="w"> </span>
<span class="w">                                     </span><span class="n">verbose</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span>
</code></pre></div>
<h3 id="expression-scaling">Expression scaling</h3>
<p>PCA is highly affected by the scale of values in our data. Accordingly, appropriate application of PCA requires that expression values are scaled to mean = 0, and standard deviation = 1. We can perform this scaling using the <code>ScaleData()</code> <code>Seurat</code> function.</p>
<p><div class="highlight"><pre><span></span><code><span class="n">norm_seurat</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ScaleData</span><span class="p">(</span><span class="n">norm_seurat</span><span class="p">)</span>
</code></pre></div>
Following <code>ScaleData()</code>, the <code>scale.data</code> "slot" of our <code>RNA</code> assay is not populated. Below we visualize CD63 expression of all three of out expression data slots.</p>
<p><img alt="" src="../images/NormExp2.png" /></p>
<p>Now that we have our scaled data comprising 2,000 genes, we can perform PCA, using the <code>RunPCA()</code> <code>Seurat</code> function.</p>
<div class="highlight"><pre><span></span><code><span class="n">norm_seurat</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">RunPCA</span><span class="p">(</span><span class="n">norm_seurat</span><span class="p">)</span>
</code></pre></div>
<p>Next, we can visualize our first two principal components using the <code>PCAPlot()</code> <code>Seurat</code> function. Additionally, we will color our points by samples, "ctrl" and "stim", to evaluate whether variability in our data reflects these differences.</p>
<div class="highlight"><pre><span></span><code><span class="n">pPlot</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">PCAPlot</span><span class="p">(</span><span class="n">norm_seurat</span><span class="p">,</span><span class="w"> </span><span class="n">group.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;sample&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">ggtitle</span><span class="p">(</span><span class="s">&quot;PCA&quot;</span><span class="p">,</span>
<span class="w">          </span><span class="n">subtitle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;RNA Assay&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme</span><span class="p">(</span>
<span class="w">    </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bottom&quot;</span>
<span class="w">  </span><span class="p">)</span>

<span class="n">pPlot</span>
</code></pre></div>
<p><img alt="" src="images/DimRed1.png" /></p>
<h3 id="uniform-manifold-approximation">Uniform manifold approximation</h3>
<p>We just visualized the first two principal components in our data. However, it is more than likely that important variability in our data is captured by additional principal components. Uniform manifold approximation (UMAP) is a procedure for visualizing many variables in two dimensions. Importantly, UMAP should only be considered as a visualization tool, and our PCA matrix is what is leveraged in downstream clustering analyses. Next, we'll run UMAP with the <code>RunUMAP()</code> <code>Seurat</code> function, on the first 10 principal components using the <code>dims</code> argument.</p>
<div class="highlight"><pre><span></span><code><span class="n">norm_seurat</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">RunUMAP</span><span class="p">(</span><span class="n">norm_seurat</span><span class="p">,</span><span class="w"> </span><span class="n">dims</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">seq</span><span class="p">(</span><span class="m">10</span><span class="p">))</span>
</code></pre></div>
<p>Now we can visualize our UMAP output side-by-side with our first two principal components.</p>
<div class="highlight"><pre><span></span><code><span class="n">uPlot</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">UMAPPlot</span><span class="p">(</span><span class="n">norm_seurat</span><span class="p">,</span><span class="w"> </span><span class="n">group.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;sample&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">ggtitle</span><span class="p">(</span><span class="s">&quot;UMAP&quot;</span><span class="p">,</span>
<span class="w">          </span><span class="n">subtitle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;RNA Assay&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme</span><span class="p">(</span>
<span class="w">    </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bottom&quot;</span>
<span class="w">  </span><span class="p">)</span>

<span class="nf">plot_grid</span><span class="p">(</span><span class="n">pPlot</span><span class="p">,</span><span class="w"> </span><span class="n">uPlot</span><span class="p">,</span><span class="w"> </span><span class="n">nrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">)</span>
</code></pre></div>
<p><img alt="" src="images/DimRed2.png" /></p>
<h2 id="alternative-expression-normalization-and-scaling-with-sctransform">Alternative expression normalization and scaling with <code>SCTransform()</code></h2>
<p>In the <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1874-1">Hafemeister and Satija, 2019</a> paper the authors explored the issues with simple transformations. Specifically they evaluated the standard log normalization approach and found that genes with different abundances are affected differently and that effective normalization (using the log transform) is only observed with low/medium abundance genes (Figure 1D, below). Additionally, substantial imbalances in variance were observed with the log-normalized data (Figure 1E, below). In particular, cells with low total UMI counts exhibited disproportionately higher variance for high-abundance genes, dampening the variance contribution from other gene abundances.</p>
<p><img alt="" src="../images/satFig.png" /></p>
<p>The proposed solution was the use of <strong>Regularized Negative Binomial Regression</strong>, as implemented in the <code>Seurat</code> <code>SCTransform()</code> function. This approach</p>
<ul>
<li>Does not assume fixed number of UMI counts per cell</li>
<li>Normalizes expression values based on their Pearson residuals from negative binomial regression of UMI counts of each gene against total UMI counts of each profile.</li>
</ul>
<p>In the next code chunk, we demonstrate usage of the <code>SCTransform()</code> function. <strong>DO NOT RUN THIS.</strong> <code>SCTransform()</code> is more computationally expensive and takes several minutes to run. Instead we will read in a a saved <code>norm_seurat</code> object, for which <code>SCTransform()</code> has been performed.</p>
<div class="highlight"><pre><span></span><code>### DON&#39;T RUN!!
norm_seurat &lt;- SCTransform(norm_seurat)
</code></pre></div>
<h4 id="read-in-sctransorm-seurat-object">Read in <code>SCTransorm()</code> <code>Seurat</code> object</h4>
<div class="highlight"><pre><span></span><code>norm_seurat &lt;- readRDS(file.path(baseDir, &quot;data/normalized_seurat.rds&quot;))
</code></pre></div>
<p>Rather than replacing the data in our <code>RNA</code> assay, <code>SCTransform()</code> creates a new assay <code>SCT</code>, and sets it as the default assay. This new assay contains fully populated count, normalized, and scaled data, in the <code>count</code>, <code>data</code>, and <code>scale.data</code> slots, respectively. </p>
<div class="highlight"><pre><span></span><code><span class="n">norm_seurat</span>
</code></pre></div>
<p><img alt="" src="../images/sct1.png" /></p>
<div class="highlight"><pre><span></span><code><span class="nf">DefaultAssay</span><span class="p">(</span><span class="n">norm_seurat</span><span class="p">)</span>
</code></pre></div>
<p><img alt="" src="../images/sct2.png" /></p>
<p>Importantly, <code>Seurat</code> functions will generally use the active (i.e. default) assay unless otherwise specified. Since <code>SCT</code> is now our e can re-run our PCA and UMAP on the <code>SCTransform()</code> scaled data using the same commands. Importantly, this <strong>will</strong> overwrite these matrices in our <code>Seurat</code> object because these are stored separately from assays. </p>
<div class="highlight"><pre><span></span><code>## Run PCA on SCT assay
norm_seurat &lt;- RunPCA(norm_seurat)

## Run UMAP on first 10 PCs from updated PCA matrix
norm_seurat &lt;- RunUMAP(norm_seurat, dims = seq(10))
</code></pre></div>
<p>Next, we will visualize our PCA and UMAP results calculated from our <code>SCT</code> assay side-by-side with those calculated from our <code>RNA</code> assay.</p>
<div class="highlight"><pre><span></span><code><span class="c1">## Plot SCT PCA results</span>
<span class="n">pPlotSCT</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">PCAPlot</span><span class="p">(</span><span class="n">norm_seurat</span><span class="p">,</span><span class="w"> </span><span class="n">group.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;sample&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">ggtitle</span><span class="p">(</span><span class="s">&quot;PCA&quot;</span><span class="p">,</span>
<span class="w">          </span><span class="n">subtitle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;SCT Assay&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme</span><span class="p">(</span>
<span class="w">    </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bottom&quot;</span>
<span class="w">  </span><span class="p">)</span>

<span class="c1">## Plot SCT UMAP results</span>
<span class="n">uPlotSCT</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">UMAPPlot</span><span class="p">(</span><span class="n">norm_seurat</span><span class="p">,</span><span class="w"> </span><span class="n">group.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;sample&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">ggtitle</span><span class="p">(</span><span class="s">&quot;UMAP&quot;</span><span class="p">,</span>
<span class="w">          </span><span class="n">subtitle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;SCT Assay&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme</span><span class="p">(</span>
<span class="w">    </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bottom&quot;</span>
<span class="w">  </span><span class="p">)</span>

<span class="c1"># Combine Plots</span>
<span class="nf">plot_grid</span><span class="p">(</span><span class="n">pPlot</span><span class="p">,</span><span class="w"> </span><span class="n">pPlotSCT</span><span class="p">,</span><span class="w"> </span>
<span class="w">          </span><span class="n">uPlot</span><span class="p">,</span><span class="w"> </span><span class="n">uPlotSCT</span><span class="p">,</span><span class="w"> </span><span class="n">nrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span>
</code></pre></div>
<p><img alt="" src="../images/dimRed3.png" /></p>
<p>Regardless of our normalization approach, it is clear that much of the variability in our data arise from differences between the data sets. This may arise from various sources. For one, our two samples represent two different conditions, i.e. control and interferon beta-treatment. Alternatively, these difference can arise from numerous technical factors during sample collection and library preparation, akin to "batch" effects, which are common in RNAseq experiments. </p>
<p>Importantly, we expect to find most if not all of the same cell types represented these samples/conditions, and we will seek to characterize these cell types via clustering analysis. Thus, we'd like cells of the same type to cluster together in order to make these results more interpretable and cell types easier to characterize. </p>
<p>Next, we will perform data integration to "harmonize" the expression profiles across these samples with the goal of overlaying transcriptomic profiles of shared cell types.</p>
<h2 id="data-integration">Data Integration</h2>
<p>The data integration workflow from the <code>Seurat</code> package is carried out in four main steps</p>
<ol>
<li>Identification of shared highly variable genes across data sets, performed by the <code>SelectIntegrationFeatures()</code>.</li>
<li>Ensure that each data set has been normalized and scaled based on these features, performed by <code>PrepSCTIntegration()</code>.</li>
<li>Scores the relative similarity of individual cells across data sets, performed by <code>FindIntegrationAnchors()</code>. This is actually performed in two sub-steps:</li>
<li>Canonical Correlation Analysis (CCA)</li>
<li>Mutual Nearest Neighbors (MMN)</li>
<li>Projects the expression profiles of individual cells to be more similar to their counterparts across data sets, performed by <code>IntegrateData()</code>.</li>
</ol>
<h4 id="canonical-correlation-analysis-cca">Canonical Correlation Analysis (CCA)</h4>
<p>The implementation of CCA in <code>Seurat</code> is effectively very similar to PCA. Whereas with PCA we seek to capture the majority of variability in a single data set in in a few features, with CCA we seek to capture the majority of "shared" variability between data sets in fewer features. Thus after implementing CCA we should obtain a set of features with reduced sample-specific variability that should capture shared signal of cell types represented across samples.</p>
<h4 id="mutual-nearest-neighbors-mnn">Mutual Nearest Neighbors (MNN)</h4>
<p>MNN comprises two main steps. First, for each profile we identify its "K" most similar cell profiles ("nearest neighbors") in the other data set based on their CCA features. Second, "Mutual nearest neighbors" are instances in which pairs of inter-data set cell profiles are complementary identified as nearest neighbors of one another. When we perform integration, the expression of each cell profile in one data set will be projected to be more similar to its MNN counterparts in the other data set.</p>
<p>We begin by splitting our <code>Seurat</code> object into a <code>list</code> object of two <code>Seurat</code> objects for each sample, "ctrl" and "stim", with the <code>SplitObject()</code> <code>Seurat</code> function.</p>
<div class="highlight"><pre><span></span><code><span class="n">split_seurat</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">SplitObject</span><span class="p">(</span><span class="n">norm_seurat</span><span class="p">,</span><span class="w"> </span><span class="n">split.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;sample&quot;</span><span class="p">)</span>

<span class="n">split_seurat</span>
</code></pre></div>
<p><img alt="" src="../images/splitSeurat.png" /></p>
<p>At this point we will not be doing anything with our <code>norm_seurat</code> object. So let's remove it to save space.</p>
<div class="highlight"><pre><span></span><code><span class="nf">rm</span><span class="p">(</span><span class="n">norm_seurat</span><span class="p">)</span>
<span class="nf">gc</span><span class="p">()</span>
</code></pre></div>
<p>**Sadly, at this point the following integration steps are quite memory expensive and slow. For this reason we give an example of how to run the integration workflow, but we will skip running the code. Instead will read in a previously integrated <code>Seurat</code> object generated by these steps.</p>
<h3 id="select-integration-features">Select integration features</h3>
<div class="highlight"><pre><span></span><code><span class="c1">## DON&#39;T RUN!</span>
<span class="n">integ_features</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">SelectIntegrationFeatures</span><span class="p">(</span><span class="n">object.list</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">split_seurat</span><span class="p">,</span><span class="w"> </span>
<span class="w">                                            </span><span class="n">nfeatures</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3000</span><span class="p">)</span><span class="w"> </span>
</code></pre></div>
<h3 id="prepare-the-sct-list-object-for-integration">Prepare the SCT list object for integration</h3>
<div class="highlight"><pre><span></span><code><span class="c1">## DON&#39;T RUN!</span>
<span class="n">split_seurat</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">PrepSCTIntegration</span><span class="p">(</span><span class="n">object.list</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">split_seurat</span><span class="p">,</span><span class="w"> </span>
<span class="w">                                   </span><span class="n">anchor.features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">integ_features</span><span class="p">)</span>
</code></pre></div>
<h3 id="perform-cca-and-identify-mutual-nearest-neighbots">Perform CCA and identify mutual nearest neighbots</h3>
<div class="highlight"><pre><span></span><code><span class="c1">## DON&#39;T RUN!</span>
<span class="n">integ_anchors</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">FindIntegrationAnchors</span><span class="p">(</span><span class="n">object.list</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">split_seurat</span><span class="p">,</span><span class="w"> </span>
<span class="w">                                        </span><span class="n">normalization.method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;SCT&quot;</span><span class="p">,</span><span class="w"> </span>
<span class="w">                                        </span><span class="n">anchor.features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">integ_features</span><span class="p">)</span>
</code></pre></div>
<h3 id="integrate-data">Integrate data</h3>
<div class="highlight"><pre><span></span><code><span class="c1">## DON&#39;T RUN!</span>
<span class="n">integ_seurat</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">IntegrateData</span><span class="p">(</span><span class="n">anchorset</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">integ_anchors</span><span class="p">,</span><span class="w"> </span>
<span class="w">                                   </span><span class="n">normalization.method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;SCT&quot;</span><span class="p">)</span>
</code></pre></div>
<h4 id="read-in-integrated-seurat-object">Read in integrated <code>Seurat</code> object</h4>
<div class="highlight"><pre><span></span><code><span class="n">integ_seurat</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">readRDS</span><span class="p">(</span><span class="nf">file.path</span><span class="p">(</span><span class="n">baseDir</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;data/integrated_seurat.rds&quot;</span><span class="p">))</span>
</code></pre></div>
<p>Similarly to when we ran <code>SCTransform()</code>, the integration workflow results in new assay in our <code>Seurat</code> object, <code>integrated</code> and set it as the active (default) assay.</p>
<div class="highlight"><pre><span></span><code><span class="n">integ_seurat</span>
</code></pre></div>
<p><img alt="" src="../images/integ1.png" /></p>
<div class="highlight"><pre><span></span><code><span class="nf">DefaultAssay</span><span class="p">(</span><span class="n">integ_seurat</span><span class="p">)</span>
</code></pre></div>
<p><img alt="" src="../images/integ2.png" /></p>
<h3 id="check-integrated-data-set">Check integrated data set</h3>
<p>Now that we have performed the integration procedure, we can interrogate our <code>integrated</code> assay for differences between samples using dimensionality reduction.</p>
<p>As before we'll re-run PCA, followed by re-running UMAP on the first 10 principal components.</p>
<div class="highlight"><pre><span></span><code>## Run PCA on integrated assay
integ_seurat &lt;- RunPCA(integ_seurat)

## Run UMAP on first 10 PCs from updated PCA matrix
integ_seurat &lt;- RunUMAP(integ_seurat, dims = seq(10))
</code></pre></div>
<p>Finally, we will visualize our PCA and UMAP results calculated from our <code>integrated</code> assay side-by-side with those calculated from the <code>SCT</code> assay of the un-integrated data.</p>
<div class="highlight"><pre><span></span><code><span class="c1">## Plot integrated PCA results</span>
<span class="n">pPlotInt</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">PCAPlot</span><span class="p">(</span><span class="n">integ_seurat</span><span class="p">,</span><span class="w"> </span><span class="n">group.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;sample&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">ggtitle</span><span class="p">(</span><span class="s">&quot;PCA&quot;</span><span class="p">,</span>
<span class="w">          </span><span class="n">subtitle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Integrated Assay&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme</span><span class="p">(</span>
<span class="w">    </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bottom&quot;</span>
<span class="w">  </span><span class="p">)</span>

<span class="c1">## Plot integrated UMAP results</span>
<span class="n">uPlotInt</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">UMAPPlot</span><span class="p">(</span><span class="n">integ_seurat</span><span class="p">,</span><span class="w"> </span><span class="n">group.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;sample&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">ggtitle</span><span class="p">(</span><span class="s">&quot;UMAP&quot;</span><span class="p">,</span>
<span class="w">          </span><span class="n">subtitle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Integrated Assay&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme</span><span class="p">(</span>
<span class="w">    </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bottom&quot;</span>
<span class="w">  </span><span class="p">)</span>

<span class="c1"># Combine Plots</span>
<span class="nf">plot_grid</span><span class="p">(</span><span class="n">pPlotSCT</span><span class="p">,</span><span class="w"> </span><span class="n">pPlotInt</span><span class="p">,</span><span class="w"> </span>
<span class="w">          </span><span class="n">uPlotSCT</span><span class="p">,</span><span class="w"> </span><span class="n">uPlotInt</span><span class="p">,</span><span class="w"> </span><span class="n">nrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span>
</code></pre></div>
<p><img alt="" src="../images/dimRed4.png" /></p>
<p>That's it! Now, we have a fully processed and integrated data set. In the next section we will go over procedures for clustering scRNAseq to facilitate cell population identification.</p>
<p>You can now open "05_clustering.Rmd" to continue on the the next section.</p>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../03_quality_control/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Quality Control">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                Quality Control
              </div>
            </div>
          </a>
        
        
          
          <a href="../05_clustering/" class="md-footer__link md-footer__link--next" aria-label="Next: Clustering">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                Clustering
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tabs", "navigation.tabs.sticky", "navigation.sections", "content.action.edit", "navigation.footer", "content.code.copy"], "search": "../../assets/javascripts/workers/search.f886a092.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.aecac24b.min.js"></script>
      
        <script src="../../javascripts/mathjax.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
        <script src="https://unpkg.com/tablesort@5.3.0/dist/tablesort.min.js"></script>
      
        <script src="../../javascripts/tablesort.js"></script>
      
    
  </body>
</html>